library(jsonlite)
library(R.utils)
library(data.table)
library(stringr)
library(dplyr)
library(openxlsx)
library(ggplot2)

##NOTES
runFisherTest <- function(signifRegions,nonSignifRegions,RegionsInList){
  signifInGroup = length(intersect(signifRegions, RegionsInList))
  signifNotInGroup = length(setdiff(signifRegions, RegionsInList))
  nonSignifInGroup = length(intersect(nonSignifRegions, RegionsInList))
  nonSignifNotInGroup = length(setdiff(nonSignifRegions, RegionsInList))
  result = fisher.test(matrix(c(signifInGroup, signifNotInGroup, nonSignifInGroup,  nonSignifNotInGroup), 2, 2), alternative="greater")
  return(result$p.value)
}

###Get the variables we need
library(config)
Sys.setenv(R_CONFIG_ACTIVE = "300bcg")
config <- config::get()
for (x in names(config)){eval(parse(text=paste0(x,"='",config[[x]],"'")))}

folderToSaveOverlapResults = file.path(data_dir_300BCG,"atacSeq","trainedImmGeneSets_overlapResults")
mkdirs(folderToSaveOverlapResults)

folderToSaveImages = file.path(data_dir_300BCG,"atacSeq","trainedImmGeneSets_overlapResults","figures")
mkdirs(folderToSaveImages)

###Main ATAC results (Non-resp vs Resp, D14 vs D0 and D90 vs D0)
dirWithAtacResults = file.path(data_dir_300BCG, "atacSeq","trainedImmGeneSets","300BCG_results") #RESULTS GENERATED BY LUKAS
resultFiles = list.files(dirWithAtacResults,pattern = '^TRIM.*')
comparisonsToCheck = gsub('^TRIM\\.(.*)\\.csv$','\\1',resultFiles)

#PC2 ATAC results (Chromatin PC2, Fig 6D)
resultFiles2 = list.files(dirWithAtacResults,pattern = '^PC2.*')
comparisonsToCheck2 = gsub('\\.csv','',resultFiles2)

#Both the above two results (files) look something like:
#,hg38_region,PC2,TSS_PROXIMAL:gene_ensembl,TSS_PROXIMAL:gene_name,GENE_AND_DISTAL_10kb:gene_ensembl,GENE_AND_DISTAL_10kb:gene_name
#CONS00000000000,chr1:9832:10510,4.268472173211968e-05,,,,
#CONS00000000002,chr1:180528:181792,-0.00013434493210416604,,,,
#CONS00000000004,chr1:191093:192116,0.0008125838668124572,,,,
#CONS00000000010,chr1:629677:631001,-0.0014203753041459378,,,,
#CONS00000000013,chr1:633765:634673,-0.002436686593478521,,,,
#CONS00000000014,chr1:778383:779427,0.003997069086116521,,,,
#CONS00000000017,chr1:826792:827945,0.003594853605627468,,,,
#CONS00000000023,chr1:869511:870386,0.0022216547237316284,,,,
#CONS00000000026,chr1:904209:905698,0.0038691278274166773,,,,
#CONS00000000027,chr1:906565:907257,0.003418114734795306,,,,
#CONS00000000028,chr1:920906:921505,0.0023556815213879915,,,,
#CONS00000000029,chr1:923387:924216,0.001136183853252718,,,ENSG00000187634,SAMD11
#CONS00000000032,chr1:940183:942132,0.002762648121122749,,,ENSG00000187634,SAMD11
#...

resultFilesFull = c(resultFiles,resultFiles2)
comparisonsToCheckFull = c(comparisonsToCheck,comparisonsToCheck2)

###Processed public data
dirWithPublicResults = file.path(data_dir_300BCG, "atacSeq", "trainedImmGeneSets")
publicMetadata = fread(file = file.path(dirWithPublicResults,'processed','metadata_trainedImmunityGenesRegions.csv'),data.table = F)
publicMetadata$'public_data_folder' = file.path(dirWithPublicResults,'processed','converted_r97')

addScData = F
if (addScData){
  ###Now add the single cell results in the same framework as the public TRIM data. This means the SC data also serves as a reference (this is the result of an analysis in src_lukas)
  singleCellFiles = list.files(dirWithAtacResults,pattern = '^scRNAseq.scLM.*')
  nrowVec = c()
  for (fileSel in singleCellFiles){
    scData = fread(file = file.path(dirWithAtacResults,singleCellFiles[[1]]))
    nrowVec = c(nrowVec,nrow(scData))
  }
  
  singleCellComparisons = gsub('^scRNAseq\\.(.*)\\.csv','\\1',singleCellFiles)
  scMetadata = as.data.frame(matrix(NA,ncol=length(publicMetadata),nrow = length(singleCellComparisons)))
  colnames(scMetadata) = colnames(publicMetadata)
  scMetadata$filename = gsub('\\.csv$','',singleCellFiles)
  scMetadata$assay = "scRNA-seq" 
  scMetadata$direction = "both" 
  scMetadata$PMID = "privateScData"
  scMetadata$comparison = singleCellComparisons
  scMetadata$cell_type = "PBMC"
  scMetadata$orig_sheet = "not applicable"
  scMetadata$full_result = "yes"
  scMetadata$description = singleCellComparisons
  scMetadata$orig_columns = "human_gene_ensembl"
  scMetadata$data_type = "gene scores"
  scMetadata$organism = "hs"
  scMetadata$region_filter = "enhancers"
  scMetadata[grep('TSS_PROXIMAL',singleCellComparisons),'region_filter'] = "promoters"
  scMetadata$orig_filename = singleCellFiles
  scMetadata$n_rows = nrowVec
  scMetadata$'public_data_folder' = dirWithAtacResults
}

runComparisonTests <- function(filenameSel,
                               atacResults,
                               publicResults,
                               dataTypeSel,
                               full_result_logical,
                               region_filter,
                               direction,
                               topComparisons= c('top1000','top2000','p_0_05','padj_0_1'),
                               testTypes = c('fisher','GSEA'),
                               comparisonSel,
                               fgseaPerm=10^5){
  
  ###"gene list","gene scores","region list","region scores"
  ###If full_result --> universe = overlap
  ###If not full_result --> universe = our regions
  ###Universe can be distal or promoter
  publicResults = publicResults[order(publicResults$pval, decreasing = F),]
  
  resultsOutputList = list()
  regionUniverses = c('distal','promoter')
  if (region_filter =="promoters"){
    regionUniverses = c('promoter')
  }else if (region_filter =="enhancers"){
    regionUniverses = c('distal')
  }
  
  for (regionUniverseSel in regionUniverses){
    resultsOutputList[[regionUniverseSel]] = list()
    
    #We filter our data for a certain universe, which can be all promoter associated regions
    #or all distal regions.
    #The initial assignment of a region to a gene (each region is assigned to exactly one gene) was done by UROPA
    #So, subset to the region of interest
    if (regionUniverseSel=='promoter'){
      atacResults.filt = atacResults[which(atacResults$gene_mapping==regionUniverseSel),]
    }else{
      atacResults.filt = atacResults[which(atacResults$gene_mapping %in% c('distal','promoter')),]
    }
    
    ##Determine background/universe
    #If there is not a full public result, take our data as the universe
    universeRegions = atacResults.filt$peak_id
    if (full_result_logical){
      #If there is a full public result, take the overlap as a universe
      allRegionsPublic = publicResults$`300BCG_region`
      allRegionsPublic = allRegionsPublic[nchar(allRegionsPublic)>0]
      allRegionsPublic = str_split(allRegionsPublic, pattern = ';')
      allRegionsPublic = unique(unlist(allRegionsPublic))
      universeRegions = intersect(universeRegions,allRegionsPublic)
      atacResults.filt = atacResults.filt[which(atacResults.filt$peak_id %in% universeRegions),]
    }
    atacResults.filt = atacResults.filt[order(atacResults.filt$ranks, decreasing = F),]
    
    ####Separate Up and Down and both
    if (direction=="both" & dataTypeSel %in% c("region scores","gene scores")){
      directionsToCheck = c('both','down','up')
    }else{
      directionsToCheck = c(direction,'both')
    }
    
    for (directionSel in directionsToCheck){
      print(directionSel)
      
      resultsOutputList[[regionUniverseSel]][[directionSel]] = list()
      
      #Filter for direction our data for Fisher test
      if ((!('LFC' %in% colnames(atacResults))) & ('PC2' %in% colnames(atacResults))){
        if (directionSel=='up'){
          atacResults.filt.dir = atacResults.filt[which(atacResults.filt$PC2>=0),,drop=F]
        }else if (directionSel=='down'){
          atacResults.filt.dir = atacResults.filt[which(atacResults.filt$PC2<0),,drop=F]
        }else{
          atacResults.filt.dir = atacResults.filt
        }
        
        ##Re-sort
        atacResults.filt.dir = atacResults.filt.dir[order(abs(atacResults.filt.dir$PC2), decreasing = T),]
      }else{
        if (directionSel=='up'){
          atacResults.filt.dir = atacResults.filt[which(atacResults.filt$LFC>=0),,drop=F]
        }else if (directionSel=='down'){
          atacResults.filt.dir = atacResults.filt[which(atacResults.filt$LFC<0),,drop=F]
        }else{
          atacResults.filt.dir = atacResults.filt
        }
        
        ##Re-sort
        atacResults.filt.dir = atacResults.filt.dir[order(atacResults.filt.dir$pval, decreasing = F),]
      }
      
      ##Set our top 1000 for Fisher
      top1000Atac = atacResults.filt.dir[c(1:1000),'peak_id']
      
      #Filter for direction public data
      publicResults.filt = publicResults
      if (dataTypeSel %in% c("gene scores","region scores")){
        if (directionSel=='up'){
          publicResults.filt = publicResults.filt[which(publicResults.filt$LFC>=0),,drop=F]
        }else if (directionSel=='down'){
          publicResults.filt = publicResults.filt[which(publicResults.filt$LFC<0),,drop=F]
        }
      }
      
      ###If we have scores and full results, we can select top 1000, 2000, and significance cutoffs,
      #If not, we select all
      if (dataTypeSel %in% c("gene scores","region scores") & full_result_logical){
        topComparisonsSel = topComparisons
      }else{
        topComparisonsSel = c("allReported")
      }
      
      for (topSel in topComparisonsSel){
        print(topSel)
        resultsOutputList[[regionUniverseSel]][[directionSel]][[topSel]] = list()
        #Filter their regions for topX or cutoff --> we basically treat theirs as a gene set, similar to e.g. a KEGG pathway
        if (topSel=="allReported"){
          allRegions = publicResults.filt$`300BCG_region`
        }else if (topSel=='p_0_05'){
          allRegions = publicResults.filt[which(publicResults.filt$pval<=0.05),"300BCG_region"]
        }else if (topSel=='padj_0_1'){
          allRegions = publicResults.filt[which(publicResults.filt$padj<=0.1),"300BCG_region"]
        }else if (startsWith(x = topSel, prefix = 'top')){
          topNr = as.numeric(gsub('top','',topSel))
          publicResultsSorted = publicResults.filt[order(publicResults.filt$pval,decreasing = F),]
          allRegions = publicResultsSorted[,"300BCG_region"]
          allRegions = unlist(allRegions)
          allRegions = allRegions[nchar(allRegions)>0]
          allRegions = str_split(allRegions, pattern = ';')
          allRegions = unlist(allRegions)
          allRegions = allRegions[!duplicated(allRegions)]
          allRegions2 = allRegions[which(allRegions %in% universeRegions)]
          allRegions = intersect(allRegions,universeRegions)
          stopifnot(all.equal(allRegions,allRegions2))
          allRegions = allRegions[c(1:topNr)]
        }
        allRegions = unlist(allRegions)
        allRegions = allRegions[nchar(allRegions)>0]
        allRegions = str_split(allRegions, pattern = ';')
        allRegions = unique(unlist(allRegions))
        selRegionsPublic = intersect(universeRegions,allRegions)
        
        if (length(selRegionsPublic)<10){next}
        
        ##FISHER AND GSEA
        for (testTypeSel in testTypes){
          print(testTypeSel)
          
          if (testTypeSel=='GSEA'){
            
            #Create a single pathway list as a pseudo-pathway database
            pathwaysListSingleEntry = list()
            pathwaysListSingleEntry[[filenameSel]] = selRegionsPublic
            
            #Use our data to calculate "ranks" (-log10(p-value) * sign(logFC))
            ranks = atacResults.filt$ranks #-log10(atacResults.filt$pval)*sign(atacResults.filt$LFC)
            names(ranks) = atacResults.filt$peak_id
            #In the case of looking at up and down at the same time, we actually want to see if there is an overlap, regardless of direction
            #So, we change the ranks to absolute values
            if (directionSel=='both'){
              ranks = abs(ranks)
            }
            ranks = sort(ranks,decreasing = F)
            library(fgsea)
            
            set.seed(42)
            fgseaRes <- fgsea(pathways = pathwaysListSingleEntry, 
                              stats    = ranks,
                              minSize  = 15,
                              maxSize  = 5000,
                              nperm=fgseaPerm,
            )
            fgseaRes
            
            ##Save the FGSEA result as an RDS file
            #ranks
            #fgseaRes
            #pathwaysListSingleEntry
            #NOTE: this is hardcoded
            saveRDS(object = fgseaRes, file = file.path(folderToSaveOverlapResults,
                                                        paste0(paste('fgseaRes',comparisonSel,filenameSel,topSel,directionSel,regionUniverseSel,sep='__'),'.RDS')))
            saveRDS(object = ranks, file = file.path(folderToSaveOverlapResults,
                                                     paste0(paste('ranks',comparisonSel,filenameSel,topSel,directionSel,regionUniverseSel,sep='__'),'.RDS')))
            saveRDS(object = pathwaysListSingleEntry, file = file.path(folderToSaveOverlapResults,
                                                                       paste0(paste('pathwaysListSingleEntry',comparisonSel,filenameSel,topSel,directionSel,regionUniverseSel,sep='__'),'.RDS')))
            
            if (length(fgseaRes$pval)==0){next}
            
            pValSel = fgseaRes$pval * sign(fgseaRes$NES)
            if (directionSel=='down'){
              pValSel = - pValSel
            }
            
            makeCustomPlot = F
            if (makeCustomPlot){
              ranksForPlot = cbind.data.frame('ranks'=ranks,'class'='none')
              ranksForPlot$class = as.character(ranksForPlot$class)
              ranksForPlot$ranks = as.numeric(ranksForPlot$ranks)
              ranksForPlot[which(rownames(ranksForPlot) %in% pathwaysListSingleEntry[[1]]),'class'] = 'selected'
              
              p = ggplot(ranksForPlot, aes(x = ranks)) +
                geom_histogram(fill = "white", colour = "black") +
                facet_grid(class ~ ., scales = "free") + ggtitle(paste0('signed pval = ', pValSel))
              folderToSaveImage = file.path(folderToSaveImages,metadataNameOtherData,"GSEA_manualPlots")
              mkdirs(folderToSaveImage)
              ggsave(p, filename = file.path(folderToSaveImage,paste0(paste('GSEA',comparisonSel,filenameSel,topSel,directionSel,regionUniverseSel,sep='__'),'.pdf')),
                     width=16/2,height=9/1.5)
            }
            
            
            plotEnrichmentRaster <- function(pathway, stats, gseaParam = 1, ticksSize = 0.05){
              library(ggrastr)
              rnk <- rank(-stats)
              ord <- order(rnk)
              statsAdj <- stats[ord]
              statsAdj <- sign(statsAdj) * (abs(statsAdj)^gseaParam)
              statsAdj <- statsAdj/max(abs(statsAdj))
              pathway <- unname(as.vector(na.omit(match(pathway, names(statsAdj)))))
              pathway <- sort(pathway)
              gseaRes <- calcGseaStat(statsAdj, selectedStats = pathway, 
                                      returnAllExtremes = TRUE)
              bottoms <- gseaRes$bottoms
              tops <- gseaRes$tops
              n <- length(statsAdj)
              xs <- as.vector(rbind(pathway - 1, pathway))
              ys <- as.vector(rbind(bottoms, tops))
              toPlot <- data.frame(x = c(0, xs, n + 1), y = c(0, ys, 0))
              diff <- (max(tops) - min(bottoms))/8
              x = y = NULL
              dpiSel=400
              g <- ggplot(toPlot, aes(x = x, y = y)) + 
                geom_hline(yintercept = max(tops), colour = "red", linetype = "dashed") + 
                geom_hline(yintercept = min(bottoms), colour = "red", linetype = "dashed") + 
                geom_hline(yintercept = 0, colour = "black") + 
                rasterise(geom_point(color = "green", size = 0.05), dpi=dpiSel) + 
                rasterise(geom_line(color = "green"), dpi=dpiSel) + 
                rasterise(geom_segment(data = data.frame(x = pathway), 
                                       mapping = aes(x = x, y = -diff/2, xend = x, yend = diff/2), size = ticksSize), dpi=dpiSel) + 
                theme_bw(base_size = 6, base_family = 'Helvetica') + 
                theme(panel.border = element_blank(), panel.grid.minor = element_blank()) + 
                theme(text=element_text(size=6,family = 'Helvetica')) +
                labs(x = "rank", y = "enrichment score") 
              
              return(g)
            }
            
            p1 = plotEnrichmentRaster(pathwaysListSingleEntry[[1]], ranks) #+ ggtitle(paste0('signed pval = ', pValSel))
            folderToSaveImage = file.path(folderToSaveImages,metadataNameOtherData,'GSEA')
            mkdirs(folderToSaveImage)
            ggsave(p1, 
                   filename = file.path(folderToSaveImage,paste0(paste('GSEA',comparisonSel,filenameSel,topSel,
                                                                       directionSel,regionUniverseSel,sep='__'),'.pdf')),
                   width=2.25*(3/4),height=1.4*(3/4))
            
          }else if (testTypeSel=='fisher'){
            
            signifInBoth = length(intersect(top1000Atac,selRegionsPublic))
            signifInFirstNotSecond = length(setdiff(top1000Atac,selRegionsPublic))
            signifInSecondNotFirst = length(setdiff(selRegionsPublic,top1000Atac))
            signifInNone = length(setdiff(universeRegions,c(top1000Atac,selRegionsPublic)))
            
            pValSel = fisher.test(matrix(c(signifInBoth, signifInFirstNotSecond, signifInSecondNotFirst,  signifInNone), 2, 2), alternative="greater")$p.value
            
          }
          
          if (length(pValSel) == 0){
            pValSel=NA
            print('check this one')
          }else if (is.na(pValSel)){
            print('check this one')
          }
          
          resultsOutputList[[regionUniverseSel]][[directionSel]][[topSel]][[testTypeSel]] = pValSel
        }
      }
    }
    
    
  }
  return(resultsOutputList)
}

if (addScData){
  metadataList = list('publicMetadata'=publicMetadata, 'scMetadata'=scMetadata)
}else{
  metadataList = list('publicMetadata'=publicMetadata)
}

#Get the system arguments
args = base::commandArgs(trailingOnly=TRUE)
#Examples:
#args = c('publicMetadata',"d0_Resp_vs_NonR",'13')
fgseaPerm=10^6

print(paste0('arguments=',args))
print(length(args))

if (length(args)==0){
  ##In case there are three arguments we submit the job array
  print('starting first part')
  
  #Get current filename and path (not very pretty code, maybe improve at some point)
  initial.options <- commandArgs(trailingOnly = FALSE)
  file.arg.name <- "--file="
  script.name <- sub(file.arg.name, "", initial.options[grep(file.arg.name, initial.options)])
  script.basename <- system('pwd', intern = T)
  print(script.basename)
  print(script.name)
  errorDir = file.path(data_dir_300BCG,'stdOutErr')
  mkdirs(errorDir)
  
  for (metadataNameOtherData in names(metadataList)){
    metadataOtherData = metadataList[[metadataNameOtherData]]
    for (comparisonSel in comparisonsToCheckFull){
      for (otherDataRowIndex in c(1:nrow(metadataOtherData))){
        origJobNamePart =  paste0(" --job-name='",paste(metadataNameOtherData,comparisonSel,otherDataRowIndex,sep='__'),"_overlapWithPublicData'")
        origErrorFolderPart =  paste0(" --error=",  "'",file.path(errorDir,paste0('overlapWithPublicData_',paste(metadataNameOtherData,comparisonSel,otherDataRowIndex,sep='__'),".err'")))
        
        basicCommand = paste0("sbatch", 
                              origJobNamePart,
                              origErrorFolderPart,
                              " --export=scriptName='",file.path(script.name),"'")
        fullScriptPath = file.path(gsub('\\.R$','.batch',script.name))
        command = paste0(basicCommand, 
                         ",metadataNameOtherData='", metadataNameOtherData,"'",
                         ",comparisonSel='", comparisonSel,"'",
                         ",otherDataRowIndex='", otherDataRowIndex,"' ",
                         fullScriptPath)
        print(command)
        system(command)
      }
    }
  }
  
}else if(length(args)==3){
  
  metadataNameOtherData = args[[1]]
  comparisonSel = args[[2]]
  otherDataRowIndex = as.numeric(args[[3]])
  metadataOtherData = metadataList[[metadataNameOtherData]]
  resultsOutputFull = list()
  print(comparisonSel)
  
  ##Get filename
  csvFileName = resultFilesFull[[grep(pattern = paste0('.*',comparisonSel,'.*'),resultFilesFull)]]
  
  atacResults = fread(file = file.path(dirWithAtacResults,csvFileName),data.table = F)
  atacResults$gene_mapping = NA
  atacResults[nchar(atacResults$"GENE_AND_DISTAL_10kb:gene_ensembl")>0,'gene_mapping']='distal'#'GENE_AND_DISTAL_10kb'
  atacResults[nchar(atacResults$"TSS_PROXIMAL:gene_ensembl")>0,'gene_mapping']='promoter'#'TSS_PROXIMAL'
  
  #For the PC2 file does not have p-values or fold-changes, but just scores.
  if ((!('LFC' %in% colnames(atacResults))) & ('PC2' %in% colnames(atacResults))){
    atacResults$ranks = atacResults$PC2
    atacResults$peak_id = atacResults$V1
    
  }else{
    atacResults$ranks = -log10(atacResults$pval)*sign(atacResults$LFC)
    if (comparisonSel=="d0_Resp_vs_NonR"){
      ##Also flip signs of Resp_vs_NonR
      atacResults$ranks = -atacResults$ranks
      atacResults$LFC = -atacResults$LFC
      comparisonSel="d0_NonR_vs_Resp"
    }
  }
  
  print(paste(otherDataRowIndex,' out of ',nrow(metadataOtherData)))
  filenameSel = metadataOtherData[otherDataRowIndex,'filename']
  print(filenameSel)
  dataTypeSel = metadataOtherData[otherDataRowIndex,'data_type']
  n_rows = metadataOtherData[otherDataRowIndex,'n_rows']
  organism = metadataOtherData[otherDataRowIndex,'organism']
  full_result = metadataOtherData[otherDataRowIndex,'full_result']
  if (full_result=='yes'){
    full_result_logical=T
  }else{
    full_result_logical=F
  }
  direction = metadataOtherData[otherDataRowIndex,'direction']
  region_filter = metadataOtherData[otherDataRowIndex,'region_filter']
  
  if (n_rows<50){next}
  
  ##Load the data
  ##LOAD
  publicResultsFile = file.path(metadataOtherData[otherDataRowIndex,'public_data_folder'],paste0(filenameSel,'.csv'))
  publicResults = fread(publicResultsFile)
  
  ##For now flip sign of 32544459_1 --> this was reported in a different direction than originally thought,
  #so flip it if it wasn't flipped in the preprocessing (and do not flip if it was already)
  if (filenameSel=='main_32544459_1'){
    if (sign(publicResults[which(publicResults$human_gene_ensembl=='ENSG00000213809'),'LFC'])==1){
      #Flip only if not flipped in preprocessing
      publicResults$LFC = -publicResults$LFC
    }
  }
  
  resultsOutputList = runComparisonTests(filenameSel,
                                         atacResults,
                                         publicResults,
                                         dataTypeSel,
                                         full_result_logical,
                                         region_filter,
                                         direction,
                                         topComparisons= c('top1000','p_0_05','top2000','padj_0_1'),
                                         testTypes = c('GSEA','fisher'),
                                         comparisonSel=comparisonSel,
                                         fgseaPerm=fgseaPerm)
  
  saveRDS(object = resultsOutputList, file = file.path(folderToSaveOverlapResults,
                                                       paste0(comparisonSel,'__',metadataNameOtherData,"__",filenameSel,'__nPermFgsea=',fgseaPerm,".RDS")))
  
}