{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import warnings\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from statsmodels.formula.api import ols\n",
    "from statsmodels.stats.anova import anova_lm\n",
    "from matplotlib.ticker import FormatStrFormatter\n",
    "import re\n",
    "from collections import defaultdict\n",
    "from matplotlib.patches import Patch\n",
    "from itertools import product\n",
    "\n",
    "import misc\n",
    "from misc import rename_cytokine, rename_CM, rename_cytokines, rename_CMs, rename_bloods, rename_factors, rename_Y, make_annot, convert_partial_year, rename_pathways, rename_Y_vars, get_day_suffix\n",
    "\n",
    "import bcg_utils as utils\n",
    "\"\"\n",
    "%load_ext autoreload\n",
    "%autoreload 1\n",
    "%aimport misc\n",
    "%aimport bcg_utils\n",
    "%aimport bcg_colors\n",
    "\n",
    "from bcg_colors import *\n",
    "\n",
    "_ = misc.setup_plotting(style='ticks', context='notebook')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "CELLTYPE = 'PBMC'\n",
    "RANK_METRIC = 'p.value'\n",
    "TOP_N = 1000\n",
    "ENR_PVAL = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "DEBUG = False\n",
    "SAVE_EXT_TABLES = False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Suppl. Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using this peak annotation: ../data/DE/peaks_filtered_PBMC.csv.gz\n",
      "V1 CM\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Coefficient</th>\n",
       "      <th>T-value</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Promoter mapping</th>\n",
       "      <th>Distal mapping</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Region</th>\n",
       "      <th>Inflammatory marker</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">chr5_132477463_132479163</th>\n",
       "      <th>CXCL10</th>\n",
       "      <th>NaN</th>\n",
       "      <th>False</th>\n",
       "      <td>0.152104</td>\n",
       "      <td>7.462842</td>\n",
       "      <td>1.277508e-12</td>\n",
       "      <td>8.184353e-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CXCL9</th>\n",
       "      <th>NaN</th>\n",
       "      <th>False</th>\n",
       "      <td>0.119286</td>\n",
       "      <td>6.554293</td>\n",
       "      <td>2.992181e-10</td>\n",
       "      <td>1.916941e-05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr4_1187003_1187698</th>\n",
       "      <th>CXCL9</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <td>0.170786</td>\n",
       "      <td>5.887135</td>\n",
       "      <td>1.209636e-08</td>\n",
       "      <td>3.874767e-04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPON2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr18_61948784_61949609</th>\n",
       "      <th>CXCL9</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <td>0.230493</td>\n",
       "      <td>5.745277</td>\n",
       "      <td>2.561615e-08</td>\n",
       "      <td>5.078470e-04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PIGN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr20_50338395_50339509</th>\n",
       "      <th>CXCL10</th>\n",
       "      <th>NaN</th>\n",
       "      <th>False</th>\n",
       "      <td>0.176857</td>\n",
       "      <td>5.739482</td>\n",
       "      <td>2.640937e-08</td>\n",
       "      <td>8.459582e-04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                 Coefficient  \\\n",
       "Region                   Inflammatory marker Phenotype selected                \n",
       "chr5_132477463_132479163 CXCL10              NaN       False        0.152104   \n",
       "                         CXCL9               NaN       False        0.119286   \n",
       "chr4_1187003_1187698     CXCL9               NaN       True         0.170786   \n",
       "chr18_61948784_61949609  CXCL9               NaN       True         0.230493   \n",
       "chr20_50338395_50339509  CXCL10              NaN       False        0.176857   \n",
       "\n",
       "                                                                  T-value  \\\n",
       "Region                   Inflammatory marker Phenotype selected             \n",
       "chr5_132477463_132479163 CXCL10              NaN       False     7.462842   \n",
       "                         CXCL9               NaN       False     6.554293   \n",
       "chr4_1187003_1187698     CXCL9               NaN       True      5.887135   \n",
       "chr18_61948784_61949609  CXCL9               NaN       True      5.745277   \n",
       "chr20_50338395_50339509  CXCL10              NaN       False     5.739482   \n",
       "\n",
       "                                                                      P-value  \\\n",
       "Region                   Inflammatory marker Phenotype selected                 \n",
       "chr5_132477463_132479163 CXCL10              NaN       False     1.277508e-12   \n",
       "                         CXCL9               NaN       False     2.992181e-10   \n",
       "chr4_1187003_1187698     CXCL9               NaN       True      1.209636e-08   \n",
       "chr18_61948784_61949609  CXCL9               NaN       True      2.561615e-08   \n",
       "chr20_50338395_50339509  CXCL10              NaN       False     2.640937e-08   \n",
       "\n",
       "                                                                 Adjusted P-value  \\\n",
       "Region                   Inflammatory marker Phenotype selected                     \n",
       "chr5_132477463_132479163 CXCL10              NaN       False         8.184353e-08   \n",
       "                         CXCL9               NaN       False         1.916941e-05   \n",
       "chr4_1187003_1187698     CXCL9               NaN       True          3.874767e-04   \n",
       "chr18_61948784_61949609  CXCL9               NaN       True          5.078470e-04   \n",
       "chr20_50338395_50339509  CXCL10              NaN       False         8.459582e-04   \n",
       "\n",
       "                                                                Promoter mapping  \\\n",
       "Region                   Inflammatory marker Phenotype selected                    \n",
       "chr5_132477463_132479163 CXCL10              NaN       False                 NaN   \n",
       "                         CXCL9               NaN       False                 NaN   \n",
       "chr4_1187003_1187698     CXCL9               NaN       True                  NaN   \n",
       "chr18_61948784_61949609  CXCL9               NaN       True                  NaN   \n",
       "chr20_50338395_50339509  CXCL10              NaN       False                 NaN   \n",
       "\n",
       "                                                                Distal mapping  \n",
       "Region                   Inflammatory marker Phenotype selected                 \n",
       "chr5_132477463_132479163 CXCL10              NaN       False               NaN  \n",
       "                         CXCL9               NaN       False               NaN  \n",
       "chr4_1187003_1187698     CXCL9               NaN       True              SPON2  \n",
       "chr18_61948784_61949609  CXCL9               NaN       True               PIGN  \n",
       "chr20_50338395_50339509  CXCL10              NaN       False               NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library</th>\n",
       "      <th>Gene set</th>\n",
       "      <th>Inflammatory marker</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th>Association</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">KEGG_2019_Human</th>\n",
       "      <th>Herpes simplex virus 1 infection</th>\n",
       "      <th>CASP-8</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.552041</td>\n",
       "      <td>1.221551e-07</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>44/768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Autophagy</th>\n",
       "      <th>MMP-10</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.873498</td>\n",
       "      <td>7.852634e-07</td>\n",
       "      <td>0.000212</td>\n",
       "      <td>31/481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">GO_Biological_Process_2018</th>\n",
       "      <th>myeloid cell differentiation (GO:0030099)</th>\n",
       "      <th>IL-6</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>4.720455</td>\n",
       "      <td>1.276133e-07</td>\n",
       "      <td>0.000264</td>\n",
       "      <td>19/186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>erythrocyte differentiation (GO:0030218)</th>\n",
       "      <th>IL-6</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>5.386476</td>\n",
       "      <td>2.431777e-07</td>\n",
       "      <td>0.000264</td>\n",
       "      <td>16/139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>negative regulation of lymphocyte proliferation (GO:0050672)</th>\n",
       "      <th>IL-17C</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>7.360731</td>\n",
       "      <td>1.326005e-07</td>\n",
       "      <td>0.000291</td>\n",
       "      <td>13/86</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                  Odds ratio  \\\n",
       "Library                    Gene set                                           Inflammatory marker Phenotype selected Association               \n",
       "KEGG_2019_Human            Herpes simplex virus 1 infection                   CASP-8              NaN       True     Positive       2.552041   \n",
       "                           Autophagy                                          MMP-10              NaN       True     Positive       2.873498   \n",
       "GO_Biological_Process_2018 myeloid cell differentiation (GO:0030099)          IL-6                NaN       True     Positive       4.720455   \n",
       "                           erythrocyte differentiation (GO:0030218)           IL-6                NaN       True     Positive       5.386476   \n",
       "                           negative regulation of lymphocyte proliferation... IL-17C              NaN       True     Negative       7.360731   \n",
       "\n",
       "                                                                                                                                       P-value  \\\n",
       "Library                    Gene set                                           Inflammatory marker Phenotype selected Association                 \n",
       "KEGG_2019_Human            Herpes simplex virus 1 infection                   CASP-8              NaN       True     Positive     1.221551e-07   \n",
       "                           Autophagy                                          MMP-10              NaN       True     Positive     7.852634e-07   \n",
       "GO_Biological_Process_2018 myeloid cell differentiation (GO:0030099)          IL-6                NaN       True     Positive     1.276133e-07   \n",
       "                           erythrocyte differentiation (GO:0030218)           IL-6                NaN       True     Positive     2.431777e-07   \n",
       "                           negative regulation of lymphocyte proliferation... IL-17C              NaN       True     Negative     1.326005e-07   \n",
       "\n",
       "                                                                                                                                  Adjusted P-value  \\\n",
       "Library                    Gene set                                           Inflammatory marker Phenotype selected Association                     \n",
       "KEGG_2019_Human            Herpes simplex virus 1 infection                   CASP-8              NaN       True     Positive             0.000033   \n",
       "                           Autophagy                                          MMP-10              NaN       True     Positive             0.000212   \n",
       "GO_Biological_Process_2018 myeloid cell differentiation (GO:0030099)          IL-6                NaN       True     Positive             0.000264   \n",
       "                           erythrocyte differentiation (GO:0030218)           IL-6                NaN       True     Positive             0.000264   \n",
       "                           negative regulation of lymphocyte proliferation... IL-17C              NaN       True     Negative             0.000291   \n",
       "\n",
       "                                                                                                                                 Overlap  \n",
       "Library                    Gene set                                           Inflammatory marker Phenotype selected Association          \n",
       "KEGG_2019_Human            Herpes simplex virus 1 infection                   CASP-8              NaN       True     Positive     44/768  \n",
       "                           Autophagy                                          MMP-10              NaN       True     Positive     31/481  \n",
       "GO_Biological_Process_2018 myeloid cell differentiation (GO:0030099)          IL-6                NaN       True     Positive     19/186  \n",
       "                           erythrocyte differentiation (GO:0030218)           IL-6                NaN       True     Positive     16/139  \n",
       "                           negative regulation of lymphocyte proliferation... IL-17C              NaN       True     Negative      13/86  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library</th>\n",
       "      <th>Gene set</th>\n",
       "      <th>Inflammatory marker</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th>Association</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">KEGG_2019_Human</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Herpes simplex virus 1 infection</th>\n",
       "      <th>SIRT2</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>3.326372</td>\n",
       "      <td>1.971546e-17</td>\n",
       "      <td>5.283745e-15</td>\n",
       "      <td>84/406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGF-β-1</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>3.166777</td>\n",
       "      <td>6.967742e-16</td>\n",
       "      <td>1.790710e-13</td>\n",
       "      <td>81/406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Systemic lupus erythematosus</th>\n",
       "      <th>TGF-β-1</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>7.139197</td>\n",
       "      <td>2.637827e-13</td>\n",
       "      <td>3.389608e-11</td>\n",
       "      <td>29/79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Herpes simplex virus 1 infection</th>\n",
       "      <th>STAMPB</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.809011</td>\n",
       "      <td>1.604806e-12</td>\n",
       "      <td>4.252737e-10</td>\n",
       "      <td>74/406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CXCL1</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.809011</td>\n",
       "      <td>1.604806e-12</td>\n",
       "      <td>4.316929e-10</td>\n",
       "      <td>74/406</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                     Odds ratio  \\\n",
       "Library         Gene set                         Inflammatory marker Phenotype selected Association               \n",
       "KEGG_2019_Human Herpes simplex virus 1 infection SIRT2               NaN       True     Positive       3.326372   \n",
       "                                                 TGF-β-1             NaN       True     Positive       3.166777   \n",
       "                Systemic lupus erythematosus     TGF-β-1             NaN       True     Positive       7.139197   \n",
       "                Herpes simplex virus 1 infection STAMPB              NaN       True     Positive       2.809011   \n",
       "                                                 CXCL1               NaN       True     Positive       2.809011   \n",
       "\n",
       "                                                                                                          P-value  \\\n",
       "Library         Gene set                         Inflammatory marker Phenotype selected Association                 \n",
       "KEGG_2019_Human Herpes simplex virus 1 infection SIRT2               NaN       True     Positive     1.971546e-17   \n",
       "                                                 TGF-β-1             NaN       True     Positive     6.967742e-16   \n",
       "                Systemic lupus erythematosus     TGF-β-1             NaN       True     Positive     2.637827e-13   \n",
       "                Herpes simplex virus 1 infection STAMPB              NaN       True     Positive     1.604806e-12   \n",
       "                                                 CXCL1               NaN       True     Positive     1.604806e-12   \n",
       "\n",
       "                                                                                                     Adjusted P-value  \\\n",
       "Library         Gene set                         Inflammatory marker Phenotype selected Association                     \n",
       "KEGG_2019_Human Herpes simplex virus 1 infection SIRT2               NaN       True     Positive         5.283745e-15   \n",
       "                                                 TGF-β-1             NaN       True     Positive         1.790710e-13   \n",
       "                Systemic lupus erythematosus     TGF-β-1             NaN       True     Positive         3.389608e-11   \n",
       "                Herpes simplex virus 1 infection STAMPB              NaN       True     Positive         4.252737e-10   \n",
       "                                                 CXCL1               NaN       True     Positive         4.316929e-10   \n",
       "\n",
       "                                                                                                    Overlap  \n",
       "Library         Gene set                         Inflammatory marker Phenotype selected Association          \n",
       "KEGG_2019_Human Herpes simplex virus 1 infection SIRT2               NaN       True     Positive     84/406  \n",
       "                                                 TGF-β-1             NaN       True     Positive     81/406  \n",
       "                Systemic lupus erythematosus     TGF-β-1             NaN       True     Positive      29/79  \n",
       "                Herpes simplex virus 1 infection STAMPB              NaN       True     Positive     74/406  \n",
       "                                                 CXCL1               NaN       True     Positive     74/406  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library</th>\n",
       "      <th>Region set</th>\n",
       "      <th>Inflammatory marker</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th>Association</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Roadmap_Epigenomics_r9</th>\n",
       "      <th>H3K4me3</th>\n",
       "      <th>PD-L1</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>6.266783</td>\n",
       "      <td>6.029318e-159</td>\n",
       "      <td>4.220523e-158</td>\n",
       "      <td>778/23396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">HOCOMOCO_v11</th>\n",
       "      <th>ZBT14</th>\n",
       "      <th>PD-L1</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>5.897173</td>\n",
       "      <td>1.184486e-142</td>\n",
       "      <td>4.749788e-140</td>\n",
       "      <td>799/26192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KAISO</th>\n",
       "      <th>PD-L1</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>5.387864</td>\n",
       "      <td>4.821994e-124</td>\n",
       "      <td>9.668097e-122</td>\n",
       "      <td>812/28870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Roadmap_Epigenomics_r9</th>\n",
       "      <th>H3K4me3</th>\n",
       "      <th>CDCP1</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>4.769605</td>\n",
       "      <td>3.974334e-122</td>\n",
       "      <td>2.782034e-121</td>\n",
       "      <td>728/23396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HOCOMOCO_v11</th>\n",
       "      <th>ZBT14</th>\n",
       "      <th>GDNF</th>\n",
       "      <th>NaN</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>4.873232</td>\n",
       "      <td>2.407807e-119</td>\n",
       "      <td>9.655306e-117</td>\n",
       "      <td>767/26192</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                      Odds ratio  \\\n",
       "Library                Region set Inflammatory marker Phenotype selected Association               \n",
       "Roadmap_Epigenomics_r9 H3K4me3    PD-L1               NaN       True     Negative       6.266783   \n",
       "HOCOMOCO_v11           ZBT14      PD-L1               NaN       True     Negative       5.897173   \n",
       "                       KAISO      PD-L1               NaN       True     Negative       5.387864   \n",
       "Roadmap_Epigenomics_r9 H3K4me3    CDCP1               NaN       True     Positive       4.769605   \n",
       "HOCOMOCO_v11           ZBT14      GDNF                NaN       True     Positive       4.873232   \n",
       "\n",
       "                                                                                            P-value  \\\n",
       "Library                Region set Inflammatory marker Phenotype selected Association                  \n",
       "Roadmap_Epigenomics_r9 H3K4me3    PD-L1               NaN       True     Negative     6.029318e-159   \n",
       "HOCOMOCO_v11           ZBT14      PD-L1               NaN       True     Negative     1.184486e-142   \n",
       "                       KAISO      PD-L1               NaN       True     Negative     4.821994e-124   \n",
       "Roadmap_Epigenomics_r9 H3K4me3    CDCP1               NaN       True     Positive     3.974334e-122   \n",
       "HOCOMOCO_v11           ZBT14      GDNF                NaN       True     Positive     2.407807e-119   \n",
       "\n",
       "                                                                                      Adjusted P-value  \\\n",
       "Library                Region set Inflammatory marker Phenotype selected Association                     \n",
       "Roadmap_Epigenomics_r9 H3K4me3    PD-L1               NaN       True     Negative        4.220523e-158   \n",
       "HOCOMOCO_v11           ZBT14      PD-L1               NaN       True     Negative        4.749788e-140   \n",
       "                       KAISO      PD-L1               NaN       True     Negative        9.668097e-122   \n",
       "Roadmap_Epigenomics_r9 H3K4me3    CDCP1               NaN       True     Positive        2.782034e-121   \n",
       "HOCOMOCO_v11           ZBT14      GDNF                NaN       True     Positive        9.655306e-117   \n",
       "\n",
       "                                                                                        Overlap  \n",
       "Library                Region set Inflammatory marker Phenotype selected Association             \n",
       "Roadmap_Epigenomics_r9 H3K4me3    PD-L1               NaN       True     Negative     778/23396  \n",
       "HOCOMOCO_v11           ZBT14      PD-L1               NaN       True     Negative     799/26192  \n",
       "                       KAISO      PD-L1               NaN       True     Negative     812/28870  \n",
       "Roadmap_Epigenomics_r9 H3K4me3    CDCP1               NaN       True     Positive     728/23396  \n",
       "HOCOMOCO_v11           ZBT14      GDNF                NaN       True     Positive     767/26192  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adjusting alltogeter\n",
      "S03_ATACseq_d0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Fraction of variance explained (FVE) by donor ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Association of chromatin accessibility with ag...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Association of chromatin accessibility with se...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Association of chromatin accessibility with BM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Association of chromatin accessibility with or...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Association of chromatin accessibility with al...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Association of chromatin accessibility with sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Association of chromatin accessibility with in...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Fraction of variance explained (FVE) by donor ...\n",
       "d0_age                   Association of chromatin accessibility with ag...\n",
       "d0_sex                   Association of chromatin accessibility with se...\n",
       "d0_BMI                   Association of chromatin accessibility with BM...\n",
       "d0_oralContraceptives    Association of chromatin accessibility with or...\n",
       "d0_alcohol               Association of chromatin accessibility with al...\n",
       "d0_sampleCollectionTime  Association of chromatin accessibility with sa...\n",
       "d0_inflaMark_d0          Association of chromatin accessibility with in..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S04_ATACseq_d0_geneSets\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Chromatin accessibility gene set enrichment of...\n",
       "d0_age                   Chromatin accessibility gene set enrichment of...\n",
       "d0_sex                   Chromatin accessibility gene set enrichment of...\n",
       "d0_BMI                   Chromatin accessibility gene set enrichment of...\n",
       "d0_oralContraceptives    Chromatin accessibility gene set enrichment of...\n",
       "d0_alcohol               Chromatin accessibility gene set enrichment of...\n",
       "d0_sampleCollectionTime  Chromatin accessibility gene set enrichment of...\n",
       "d0_inflaMark_d0          Chromatin accessibility gene set enrichment of..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ext_S04.1_ATACseq_d0_geneSets_promoters\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Chromatin accessibility gene set enrichment of...\n",
       "d0_age                   Chromatin accessibility gene set enrichment of...\n",
       "d0_sex                   Chromatin accessibility gene set enrichment of...\n",
       "d0_BMI                   Chromatin accessibility gene set enrichment of...\n",
       "d0_oralContraceptives    Chromatin accessibility gene set enrichment of...\n",
       "d0_alcohol               Chromatin accessibility gene set enrichment of...\n",
       "d0_sampleCollectionTime  Chromatin accessibility gene set enrichment of...\n",
       "d0_inflaMark_d0          Chromatin accessibility gene set enrichment of..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ext_S04.2_ATACseq_d0_regionSets\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Chromatin accessibility region set enrichment ...\n",
       "d0_age                   Chromatin accessibility region set enrichment ...\n",
       "d0_sex                   Chromatin accessibility region set enrichment ...\n",
       "d0_BMI                   Chromatin accessibility region set enrichment ...\n",
       "d0_oralContraceptives    Chromatin accessibility region set enrichment ...\n",
       "d0_alcohol               Chromatin accessibility region set enrichment ...\n",
       "d0_sampleCollectionTime  Chromatin accessibility region set enrichment ...\n",
       "d0_inflaMark_d0          Chromatin accessibility region set enrichment ..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "V1 CYTO\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Coefficient</th>\n",
       "      <th>T-value</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Promoter mapping</th>\n",
       "      <th>Distal mapping</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Region</th>\n",
       "      <th>Cytokine</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr2_112699917_112701970</th>\n",
       "      <th>IL-12 (Mt, 24h)</th>\n",
       "      <th>innate after Mt stimulus</th>\n",
       "      <th>False</th>\n",
       "      <td>-0.132999</td>\n",
       "      <td>-5.699332</td>\n",
       "      <td>3.484961e-08</td>\n",
       "      <td>0.002233</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr17_36196216_36197668</th>\n",
       "      <th>CCL4 (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <td>0.634402</td>\n",
       "      <td>5.583952</td>\n",
       "      <td>7.146317e-08</td>\n",
       "      <td>0.004578</td>\n",
       "      <td>CCL3L1</td>\n",
       "      <td>CCL3L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr17_78412334_78413842</th>\n",
       "      <th>IL-12 (Mt, 24h)</th>\n",
       "      <th>innate after Mt stimulus</th>\n",
       "      <th>True</th>\n",
       "      <td>-0.106393</td>\n",
       "      <td>-5.391513</td>\n",
       "      <td>1.659259e-07</td>\n",
       "      <td>0.005315</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3_53146883_53147535</th>\n",
       "      <th>IL-1β (LPS, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <td>0.216271</td>\n",
       "      <td>5.169279</td>\n",
       "      <td>4.681035e-07</td>\n",
       "      <td>0.029989</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRKCD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr21_43352449_43353322</th>\n",
       "      <th>TNF-α (LPS, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>False</th>\n",
       "      <td>0.238631</td>\n",
       "      <td>5.090764</td>\n",
       "      <td>7.292569e-07</td>\n",
       "      <td>0.046720</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                             Coefficient  \\\n",
       "Region                   Cytokine         Phenotype                selected                \n",
       "chr2_112699917_112701970 IL-12 (Mt, 24h)  innate after Mt stimulus False       -0.132999   \n",
       "chr17_36196216_36197668  CCL4 (Ca, 24h)   trained immunity         True         0.634402   \n",
       "chr17_78412334_78413842  IL-12 (Mt, 24h)  innate after Mt stimulus True        -0.106393   \n",
       "chr3_53146883_53147535   IL-1β (LPS, 24h) trained immunity         True         0.216271   \n",
       "chr21_43352449_43353322  TNF-α (LPS, 24h) trained immunity         False        0.238631   \n",
       "\n",
       "                                                                              T-value  \\\n",
       "Region                   Cytokine         Phenotype                selected             \n",
       "chr2_112699917_112701970 IL-12 (Mt, 24h)  innate after Mt stimulus False    -5.699332   \n",
       "chr17_36196216_36197668  CCL4 (Ca, 24h)   trained immunity         True      5.583952   \n",
       "chr17_78412334_78413842  IL-12 (Mt, 24h)  innate after Mt stimulus True     -5.391513   \n",
       "chr3_53146883_53147535   IL-1β (LPS, 24h) trained immunity         True      5.169279   \n",
       "chr21_43352449_43353322  TNF-α (LPS, 24h) trained immunity         False     5.090764   \n",
       "\n",
       "                                                                                  P-value  \\\n",
       "Region                   Cytokine         Phenotype                selected                 \n",
       "chr2_112699917_112701970 IL-12 (Mt, 24h)  innate after Mt stimulus False     3.484961e-08   \n",
       "chr17_36196216_36197668  CCL4 (Ca, 24h)   trained immunity         True      7.146317e-08   \n",
       "chr17_78412334_78413842  IL-12 (Mt, 24h)  innate after Mt stimulus True      1.659259e-07   \n",
       "chr3_53146883_53147535   IL-1β (LPS, 24h) trained immunity         True      4.681035e-07   \n",
       "chr21_43352449_43353322  TNF-α (LPS, 24h) trained immunity         False     7.292569e-07   \n",
       "\n",
       "                                                                             Adjusted P-value  \\\n",
       "Region                   Cytokine         Phenotype                selected                     \n",
       "chr2_112699917_112701970 IL-12 (Mt, 24h)  innate after Mt stimulus False             0.002233   \n",
       "chr17_36196216_36197668  CCL4 (Ca, 24h)   trained immunity         True              0.004578   \n",
       "chr17_78412334_78413842  IL-12 (Mt, 24h)  innate after Mt stimulus True              0.005315   \n",
       "chr3_53146883_53147535   IL-1β (LPS, 24h) trained immunity         True              0.029989   \n",
       "chr21_43352449_43353322  TNF-α (LPS, 24h) trained immunity         False             0.046720   \n",
       "\n",
       "                                                                            Promoter mapping  \\\n",
       "Region                   Cytokine         Phenotype                selected                    \n",
       "chr2_112699917_112701970 IL-12 (Mt, 24h)  innate after Mt stimulus False                 NaN   \n",
       "chr17_36196216_36197668  CCL4 (Ca, 24h)   trained immunity         True               CCL3L1   \n",
       "chr17_78412334_78413842  IL-12 (Mt, 24h)  innate after Mt stimulus True                  NaN   \n",
       "chr3_53146883_53147535   IL-1β (LPS, 24h) trained immunity         True                  NaN   \n",
       "chr21_43352449_43353322  TNF-α (LPS, 24h) trained immunity         False                 NaN   \n",
       "\n",
       "                                                                            Distal mapping  \n",
       "Region                   Cytokine         Phenotype                selected                 \n",
       "chr2_112699917_112701970 IL-12 (Mt, 24h)  innate after Mt stimulus False               NaN  \n",
       "chr17_36196216_36197668  CCL4 (Ca, 24h)   trained immunity         True             CCL3L1  \n",
       "chr17_78412334_78413842  IL-12 (Mt, 24h)  innate after Mt stimulus True               PGS1  \n",
       "chr3_53146883_53147535   IL-1β (LPS, 24h) trained immunity         True              PRKCD  \n",
       "chr21_43352449_43353322  TNF-α (LPS, 24h) trained immunity         False               NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library</th>\n",
       "      <th>Gene set</th>\n",
       "      <th>Cytokine</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th>Association</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">GO_Biological_Process_2018</th>\n",
       "      <th>neutrophil mediated immunity (GO:0002446)</th>\n",
       "      <th>IL-6 (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.648734</td>\n",
       "      <td>6.202358e-14</td>\n",
       "      <td>7.368796e-11</td>\n",
       "      <td>86/1488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neutrophil activation involved in immune response (GO:0002283)</th>\n",
       "      <th>IL-6 (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.644827</td>\n",
       "      <td>6.678826e-14</td>\n",
       "      <td>7.368796e-11</td>\n",
       "      <td>86/1490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neutrophil degranulation (GO:0043312)</th>\n",
       "      <th>IL-6 (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.640418</td>\n",
       "      <td>9.899861e-14</td>\n",
       "      <td>7.368796e-11</td>\n",
       "      <td>85/1474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein targeting to ER (GO:0045047)</th>\n",
       "      <th>TNF-α (LPS, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>6.400963</td>\n",
       "      <td>7.694603e-08</td>\n",
       "      <td>1.728208e-04</td>\n",
       "      <td>15/112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neutrophil mediated immunity (GO:0002446)</th>\n",
       "      <th>IL-1Ra (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.060459</td>\n",
       "      <td>1.600209e-07</td>\n",
       "      <td>2.597821e-04</td>\n",
       "      <td>69/1488</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                      Odds ratio  \\\n",
       "Library                    Gene set                                           Cytokine         Phenotype        selected Association               \n",
       "GO_Biological_Process_2018 neutrophil mediated immunity (GO:0002446)          IL-6 (Ca, 24h)   trained immunity True     Positive       2.648734   \n",
       "                           neutrophil activation involved in immune respon... IL-6 (Ca, 24h)   trained immunity True     Positive       2.644827   \n",
       "                           neutrophil degranulation (GO:0043312)              IL-6 (Ca, 24h)   trained immunity True     Positive       2.640418   \n",
       "                           protein targeting to ER (GO:0045047)               TNF-α (LPS, 24h) trained immunity True     Positive       6.400963   \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-1Ra (Ca, 24h) trained immunity True     Positive       2.060459   \n",
       "\n",
       "                                                                                                                                           P-value  \\\n",
       "Library                    Gene set                                           Cytokine         Phenotype        selected Association                 \n",
       "GO_Biological_Process_2018 neutrophil mediated immunity (GO:0002446)          IL-6 (Ca, 24h)   trained immunity True     Positive     6.202358e-14   \n",
       "                           neutrophil activation involved in immune respon... IL-6 (Ca, 24h)   trained immunity True     Positive     6.678826e-14   \n",
       "                           neutrophil degranulation (GO:0043312)              IL-6 (Ca, 24h)   trained immunity True     Positive     9.899861e-14   \n",
       "                           protein targeting to ER (GO:0045047)               TNF-α (LPS, 24h) trained immunity True     Positive     7.694603e-08   \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-1Ra (Ca, 24h) trained immunity True     Positive     1.600209e-07   \n",
       "\n",
       "                                                                                                                                      Adjusted P-value  \\\n",
       "Library                    Gene set                                           Cytokine         Phenotype        selected Association                     \n",
       "GO_Biological_Process_2018 neutrophil mediated immunity (GO:0002446)          IL-6 (Ca, 24h)   trained immunity True     Positive         7.368796e-11   \n",
       "                           neutrophil activation involved in immune respon... IL-6 (Ca, 24h)   trained immunity True     Positive         7.368796e-11   \n",
       "                           neutrophil degranulation (GO:0043312)              IL-6 (Ca, 24h)   trained immunity True     Positive         7.368796e-11   \n",
       "                           protein targeting to ER (GO:0045047)               TNF-α (LPS, 24h) trained immunity True     Positive         1.728208e-04   \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-1Ra (Ca, 24h) trained immunity True     Positive         2.597821e-04   \n",
       "\n",
       "                                                                                                                                      Overlap  \n",
       "Library                    Gene set                                           Cytokine         Phenotype        selected Association           \n",
       "GO_Biological_Process_2018 neutrophil mediated immunity (GO:0002446)          IL-6 (Ca, 24h)   trained immunity True     Positive     86/1488  \n",
       "                           neutrophil activation involved in immune respon... IL-6 (Ca, 24h)   trained immunity True     Positive     86/1490  \n",
       "                           neutrophil degranulation (GO:0043312)              IL-6 (Ca, 24h)   trained immunity True     Positive     85/1474  \n",
       "                           protein targeting to ER (GO:0045047)               TNF-α (LPS, 24h) trained immunity True     Positive      15/112  \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-1Ra (Ca, 24h) trained immunity True     Positive     69/1488  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library</th>\n",
       "      <th>Gene set</th>\n",
       "      <th>Cytokine</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th>Association</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">GO_Biological_Process_2018</th>\n",
       "      <th>neutrophil activation involved in immune response (GO:0002283)</th>\n",
       "      <th>IL-12 (Mt, 24h)</th>\n",
       "      <th>innate after Mt stimulus</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>3.188731</td>\n",
       "      <td>7.146820e-17</td>\n",
       "      <td>1.561580e-13</td>\n",
       "      <td>86/430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neutrophil mediated immunity (GO:0002446)</th>\n",
       "      <th>IL-12 (Mt, 24h)</th>\n",
       "      <th>innate after Mt stimulus</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>3.061936</td>\n",
       "      <td>1.101765e-15</td>\n",
       "      <td>9.245313e-13</td>\n",
       "      <td>84/433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neutrophil degranulation (GO:0043312)</th>\n",
       "      <th>IL-12 (Mt, 24h)</th>\n",
       "      <th>innate after Mt stimulus</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>3.076635</td>\n",
       "      <td>1.269379e-15</td>\n",
       "      <td>9.245313e-13</td>\n",
       "      <td>83/426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>translation (GO:0006412)</th>\n",
       "      <th>CXCL9 (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>4.207023</td>\n",
       "      <td>2.643202e-13</td>\n",
       "      <td>5.437067e-10</td>\n",
       "      <td>46/182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>viral transcription (GO:0019083)</th>\n",
       "      <th>CXCL9 (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>6.160991</td>\n",
       "      <td>8.042083e-13</td>\n",
       "      <td>6.877918e-10</td>\n",
       "      <td>31/93</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                             Odds ratio  \\\n",
       "Library                    Gene set                                           Cytokine        Phenotype                selected Association               \n",
       "GO_Biological_Process_2018 neutrophil activation involved in immune respon... IL-12 (Mt, 24h) innate after Mt stimulus True     Negative       3.188731   \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-12 (Mt, 24h) innate after Mt stimulus True     Negative       3.061936   \n",
       "                           neutrophil degranulation (GO:0043312)              IL-12 (Mt, 24h) innate after Mt stimulus True     Negative       3.076635   \n",
       "                           translation (GO:0006412)                           CXCL9 (Ca, 24h) trained immunity         True     Negative       4.207023   \n",
       "                           viral transcription (GO:0019083)                   CXCL9 (Ca, 24h) trained immunity         True     Negative       6.160991   \n",
       "\n",
       "                                                                                                                                                  P-value  \\\n",
       "Library                    Gene set                                           Cytokine        Phenotype                selected Association                 \n",
       "GO_Biological_Process_2018 neutrophil activation involved in immune respon... IL-12 (Mt, 24h) innate after Mt stimulus True     Negative     7.146820e-17   \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-12 (Mt, 24h) innate after Mt stimulus True     Negative     1.101765e-15   \n",
       "                           neutrophil degranulation (GO:0043312)              IL-12 (Mt, 24h) innate after Mt stimulus True     Negative     1.269379e-15   \n",
       "                           translation (GO:0006412)                           CXCL9 (Ca, 24h) trained immunity         True     Negative     2.643202e-13   \n",
       "                           viral transcription (GO:0019083)                   CXCL9 (Ca, 24h) trained immunity         True     Negative     8.042083e-13   \n",
       "\n",
       "                                                                                                                                             Adjusted P-value  \\\n",
       "Library                    Gene set                                           Cytokine        Phenotype                selected Association                     \n",
       "GO_Biological_Process_2018 neutrophil activation involved in immune respon... IL-12 (Mt, 24h) innate after Mt stimulus True     Negative         1.561580e-13   \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-12 (Mt, 24h) innate after Mt stimulus True     Negative         9.245313e-13   \n",
       "                           neutrophil degranulation (GO:0043312)              IL-12 (Mt, 24h) innate after Mt stimulus True     Negative         9.245313e-13   \n",
       "                           translation (GO:0006412)                           CXCL9 (Ca, 24h) trained immunity         True     Negative         5.437067e-10   \n",
       "                           viral transcription (GO:0019083)                   CXCL9 (Ca, 24h) trained immunity         True     Negative         6.877918e-10   \n",
       "\n",
       "                                                                                                                                            Overlap  \n",
       "Library                    Gene set                                           Cytokine        Phenotype                selected Association          \n",
       "GO_Biological_Process_2018 neutrophil activation involved in immune respon... IL-12 (Mt, 24h) innate after Mt stimulus True     Negative     86/430  \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-12 (Mt, 24h) innate after Mt stimulus True     Negative     84/433  \n",
       "                           neutrophil degranulation (GO:0043312)              IL-12 (Mt, 24h) innate after Mt stimulus True     Negative     83/426  \n",
       "                           translation (GO:0006412)                           CXCL9 (Ca, 24h) trained immunity         True     Negative     46/182  \n",
       "                           viral transcription (GO:0019083)                   CXCL9 (Ca, 24h) trained immunity         True     Negative      31/93  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library</th>\n",
       "      <th>Region set</th>\n",
       "      <th>Cytokine</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th>Association</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Roadmap_Epigenomics_r9</th>\n",
       "      <th>H3K4me3</th>\n",
       "      <th>IL-6 (Sa, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>15.701918</td>\n",
       "      <td>5.170528e-275</td>\n",
       "      <td>3.619369e-274</td>\n",
       "      <td>897/23396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">HOCOMOCO_v11</th>\n",
       "      <th>ZBT14</th>\n",
       "      <th>IL-6 (Sa, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>18.720189</td>\n",
       "      <td>7.718043e-268</td>\n",
       "      <td>3.094935e-265</td>\n",
       "      <td>926/26192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KAISO</th>\n",
       "      <th>IL-6 (Sa, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>15.953653</td>\n",
       "      <td>1.441756e-231</td>\n",
       "      <td>2.890721e-229</td>\n",
       "      <td>927/28870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Roadmap_Epigenomics_r9</th>\n",
       "      <th>H3K4me3</th>\n",
       "      <th>IL-6 (Mt, 24h)</th>\n",
       "      <th>innate after Mt stimulus</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>9.086877</td>\n",
       "      <td>8.849412e-209</td>\n",
       "      <td>6.194589e-208</td>\n",
       "      <td>835/23396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HOCOMOCO_v11</th>\n",
       "      <th>MBD2</th>\n",
       "      <th>IL-6 (Sa, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>17.174464</td>\n",
       "      <td>3.040414e-205</td>\n",
       "      <td>4.064020e-203</td>\n",
       "      <td>946/32793</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                Odds ratio  \\\n",
       "Library                Region set Cytokine       Phenotype                selected Association               \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-6 (Sa, 24h) trained immunity         True     Negative      15.701918   \n",
       "HOCOMOCO_v11           ZBT14      IL-6 (Sa, 24h) trained immunity         True     Negative      18.720189   \n",
       "                       KAISO      IL-6 (Sa, 24h) trained immunity         True     Negative      15.953653   \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-6 (Mt, 24h) innate after Mt stimulus True     Negative       9.086877   \n",
       "HOCOMOCO_v11           MBD2       IL-6 (Sa, 24h) trained immunity         True     Negative      17.174464   \n",
       "\n",
       "                                                                                                      P-value  \\\n",
       "Library                Region set Cytokine       Phenotype                selected Association                  \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-6 (Sa, 24h) trained immunity         True     Negative     5.170528e-275   \n",
       "HOCOMOCO_v11           ZBT14      IL-6 (Sa, 24h) trained immunity         True     Negative     7.718043e-268   \n",
       "                       KAISO      IL-6 (Sa, 24h) trained immunity         True     Negative     1.441756e-231   \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-6 (Mt, 24h) innate after Mt stimulus True     Negative     8.849412e-209   \n",
       "HOCOMOCO_v11           MBD2       IL-6 (Sa, 24h) trained immunity         True     Negative     3.040414e-205   \n",
       "\n",
       "                                                                                                Adjusted P-value  \\\n",
       "Library                Region set Cytokine       Phenotype                selected Association                     \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-6 (Sa, 24h) trained immunity         True     Negative        3.619369e-274   \n",
       "HOCOMOCO_v11           ZBT14      IL-6 (Sa, 24h) trained immunity         True     Negative        3.094935e-265   \n",
       "                       KAISO      IL-6 (Sa, 24h) trained immunity         True     Negative        2.890721e-229   \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-6 (Mt, 24h) innate after Mt stimulus True     Negative        6.194589e-208   \n",
       "HOCOMOCO_v11           MBD2       IL-6 (Sa, 24h) trained immunity         True     Negative        4.064020e-203   \n",
       "\n",
       "                                                                                                  Overlap  \n",
       "Library                Region set Cytokine       Phenotype                selected Association             \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-6 (Sa, 24h) trained immunity         True     Negative     897/23396  \n",
       "HOCOMOCO_v11           ZBT14      IL-6 (Sa, 24h) trained immunity         True     Negative     926/26192  \n",
       "                       KAISO      IL-6 (Sa, 24h) trained immunity         True     Negative     927/28870  \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-6 (Mt, 24h) innate after Mt stimulus True     Negative     835/23396  \n",
       "HOCOMOCO_v11           MBD2       IL-6 (Sa, 24h) trained immunity         True     Negative     946/32793  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adjusting alltogeter\n",
      "S03_ATACseq_d0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Fraction of variance explained (FVE) by donor ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Association of chromatin accessibility with ag...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Association of chromatin accessibility with se...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Association of chromatin accessibility with BM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Association of chromatin accessibility with or...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Association of chromatin accessibility with al...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Association of chromatin accessibility with sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Association of chromatin accessibility with in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d0</th>\n",
       "      <td>Association of chromatin accessibility with cy...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Fraction of variance explained (FVE) by donor ...\n",
       "d0_age                   Association of chromatin accessibility with ag...\n",
       "d0_sex                   Association of chromatin accessibility with se...\n",
       "d0_BMI                   Association of chromatin accessibility with BM...\n",
       "d0_oralContraceptives    Association of chromatin accessibility with or...\n",
       "d0_alcohol               Association of chromatin accessibility with al...\n",
       "d0_sampleCollectionTime  Association of chromatin accessibility with sa...\n",
       "d0_inflaMark_d0          Association of chromatin accessibility with in...\n",
       "d0_cytokine_d0           Association of chromatin accessibility with cy..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S04_ATACseq_d0_geneSets\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Chromatin accessibility gene set enrichment of...\n",
       "d0_age                   Chromatin accessibility gene set enrichment of...\n",
       "d0_sex                   Chromatin accessibility gene set enrichment of...\n",
       "d0_BMI                   Chromatin accessibility gene set enrichment of...\n",
       "d0_oralContraceptives    Chromatin accessibility gene set enrichment of...\n",
       "d0_alcohol               Chromatin accessibility gene set enrichment of...\n",
       "d0_sampleCollectionTime  Chromatin accessibility gene set enrichment of...\n",
       "d0_inflaMark_d0          Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d0           Chromatin accessibility gene set enrichment of..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ext_S04.1_ATACseq_d0_geneSets_promoters\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Chromatin accessibility gene set enrichment of...\n",
       "d0_age                   Chromatin accessibility gene set enrichment of...\n",
       "d0_sex                   Chromatin accessibility gene set enrichment of...\n",
       "d0_BMI                   Chromatin accessibility gene set enrichment of...\n",
       "d0_oralContraceptives    Chromatin accessibility gene set enrichment of...\n",
       "d0_alcohol               Chromatin accessibility gene set enrichment of...\n",
       "d0_sampleCollectionTime  Chromatin accessibility gene set enrichment of...\n",
       "d0_inflaMark_d0          Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d0           Chromatin accessibility gene set enrichment of..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ext_S04.2_ATACseq_d0_regionSets\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d0</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Chromatin accessibility region set enrichment ...\n",
       "d0_age                   Chromatin accessibility region set enrichment ...\n",
       "d0_sex                   Chromatin accessibility region set enrichment ...\n",
       "d0_BMI                   Chromatin accessibility region set enrichment ...\n",
       "d0_oralContraceptives    Chromatin accessibility region set enrichment ...\n",
       "d0_alcohol               Chromatin accessibility region set enrichment ...\n",
       "d0_sampleCollectionTime  Chromatin accessibility region set enrichment ...\n",
       "d0_inflaMark_d0          Chromatin accessibility region set enrichment ...\n",
       "d0_cytokine_d0           Chromatin accessibility region set enrichment ..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "V2 CYTO\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Coefficient</th>\n",
       "      <th>T-value</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Promoter mapping</th>\n",
       "      <th>Distal mapping</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Region</th>\n",
       "      <th>Cytokine</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr5_150376589_150377685</th>\n",
       "      <th>GM-CSF (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <td>-0.218049</td>\n",
       "      <td>-4.993154</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.071618</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCOF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr12_116410635_116411497</th>\n",
       "      <th>IL-12 (Mt, 24h)</th>\n",
       "      <th>innate after Mt stimulus</th>\n",
       "      <th>False</th>\n",
       "      <td>0.178786</td>\n",
       "      <td>4.818798</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.078014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr7_150365781_150366490</th>\n",
       "      <th>IL-12 (Mt, 24h)</th>\n",
       "      <th>innate after Mt stimulus</th>\n",
       "      <th>False</th>\n",
       "      <td>0.143068</td>\n",
       "      <td>4.815716</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.078014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr7_117791751_117792995</th>\n",
       "      <th>IL-2RA (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <td>-0.407120</td>\n",
       "      <td>-4.783543</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.189550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTTNBP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr14_87999106_87999816</th>\n",
       "      <th>IL-17 (Ca, 7d)</th>\n",
       "      <th>heterologous immunity</th>\n",
       "      <th>True</th>\n",
       "      <td>0.144606</td>\n",
       "      <td>4.749247</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.168726</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPR65</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                              Coefficient  \\\n",
       "Region                    Cytokine         Phenotype                selected                \n",
       "chr5_150376589_150377685  GM-CSF (Ca, 24h) trained immunity         True        -0.218049   \n",
       "chr12_116410635_116411497 IL-12 (Mt, 24h)  innate after Mt stimulus False        0.178786   \n",
       "chr7_150365781_150366490  IL-12 (Mt, 24h)  innate after Mt stimulus False        0.143068   \n",
       "chr7_117791751_117792995  IL-2RA (Ca, 24h) trained immunity         True        -0.407120   \n",
       "chr14_87999106_87999816   IL-17 (Ca, 7d)   heterologous immunity    True         0.144606   \n",
       "\n",
       "                                                                               T-value  \\\n",
       "Region                    Cytokine         Phenotype                selected             \n",
       "chr5_150376589_150377685  GM-CSF (Ca, 24h) trained immunity         True     -4.993154   \n",
       "chr12_116410635_116411497 IL-12 (Mt, 24h)  innate after Mt stimulus False     4.818798   \n",
       "chr7_150365781_150366490  IL-12 (Mt, 24h)  innate after Mt stimulus False     4.815716   \n",
       "chr7_117791751_117792995  IL-2RA (Ca, 24h) trained immunity         True     -4.783543   \n",
       "chr14_87999106_87999816   IL-17 (Ca, 7d)   heterologous immunity    True      4.749247   \n",
       "\n",
       "                                                                               P-value  \\\n",
       "Region                    Cytokine         Phenotype                selected             \n",
       "chr5_150376589_150377685  GM-CSF (Ca, 24h) trained immunity         True      0.000001   \n",
       "chr12_116410635_116411497 IL-12 (Mt, 24h)  innate after Mt stimulus False     0.000003   \n",
       "chr7_150365781_150366490  IL-12 (Mt, 24h)  innate after Mt stimulus False     0.000003   \n",
       "chr7_117791751_117792995  IL-2RA (Ca, 24h) trained immunity         True      0.000003   \n",
       "chr14_87999106_87999816   IL-17 (Ca, 7d)   heterologous immunity    True      0.000004   \n",
       "\n",
       "                                                                              Adjusted P-value  \\\n",
       "Region                    Cytokine         Phenotype                selected                     \n",
       "chr5_150376589_150377685  GM-CSF (Ca, 24h) trained immunity         True              0.071618   \n",
       "chr12_116410635_116411497 IL-12 (Mt, 24h)  innate after Mt stimulus False             0.078014   \n",
       "chr7_150365781_150366490  IL-12 (Mt, 24h)  innate after Mt stimulus False             0.078014   \n",
       "chr7_117791751_117792995  IL-2RA (Ca, 24h) trained immunity         True              0.189550   \n",
       "chr14_87999106_87999816   IL-17 (Ca, 7d)   heterologous immunity    True              0.168726   \n",
       "\n",
       "                                                                             Promoter mapping  \\\n",
       "Region                    Cytokine         Phenotype                selected                    \n",
       "chr5_150376589_150377685  GM-CSF (Ca, 24h) trained immunity         True                  NaN   \n",
       "chr12_116410635_116411497 IL-12 (Mt, 24h)  innate after Mt stimulus False                 NaN   \n",
       "chr7_150365781_150366490  IL-12 (Mt, 24h)  innate after Mt stimulus False                 NaN   \n",
       "chr7_117791751_117792995  IL-2RA (Ca, 24h) trained immunity         True                  NaN   \n",
       "chr14_87999106_87999816   IL-17 (Ca, 7d)   heterologous immunity    True                  NaN   \n",
       "\n",
       "                                                                             Distal mapping  \n",
       "Region                    Cytokine         Phenotype                selected                 \n",
       "chr5_150376589_150377685  GM-CSF (Ca, 24h) trained immunity         True              TCOF1  \n",
       "chr12_116410635_116411497 IL-12 (Mt, 24h)  innate after Mt stimulus False               NaN  \n",
       "chr7_150365781_150366490  IL-12 (Mt, 24h)  innate after Mt stimulus False               NaN  \n",
       "chr7_117791751_117792995  IL-2RA (Ca, 24h) trained immunity         True            CTTNBP2  \n",
       "chr14_87999106_87999816   IL-17 (Ca, 7d)   heterologous immunity    True              GPR65  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library</th>\n",
       "      <th>Gene set</th>\n",
       "      <th>Cytokine</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th>Association</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KEGG_2019_Human</th>\n",
       "      <th>Systemic lupus erythematosus</th>\n",
       "      <th>IL-1β (Sa, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>4.925282</td>\n",
       "      <td>3.301606e-07</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>17/160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">GO_Biological_Process_2018</th>\n",
       "      <th>neutrophil degranulation (GO:0043312)</th>\n",
       "      <th>IL-2RA (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>2.081729</td>\n",
       "      <td>1.127527e-07</td>\n",
       "      <td>0.000124</td>\n",
       "      <td>69/1474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neutrophil mediated immunity (GO:0002446)</th>\n",
       "      <th>IL-2RA (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>2.060459</td>\n",
       "      <td>1.600209e-07</td>\n",
       "      <td>0.000124</td>\n",
       "      <td>69/1488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neutrophil activation involved in immune response (GO:0002283)</th>\n",
       "      <th>IL-2RA (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>2.057455</td>\n",
       "      <td>1.681391e-07</td>\n",
       "      <td>0.000124</td>\n",
       "      <td>69/1490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>antigen receptor-mediated signaling pathway (GO:0050851)</th>\n",
       "      <th>IL-1β (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.455115</td>\n",
       "      <td>5.868685e-07</td>\n",
       "      <td>0.001272</td>\n",
       "      <td>42/759</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                      Odds ratio  \\\n",
       "Library                    Gene set                                           Cytokine         Phenotype        selected Association               \n",
       "KEGG_2019_Human            Systemic lupus erythematosus                       IL-1β (Sa, 24h)  trained immunity True     Negative       4.925282   \n",
       "GO_Biological_Process_2018 neutrophil degranulation (GO:0043312)              IL-2RA (Ca, 24h) trained immunity True     Negative       2.081729   \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-2RA (Ca, 24h) trained immunity True     Negative       2.060459   \n",
       "                           neutrophil activation involved in immune respon... IL-2RA (Ca, 24h) trained immunity True     Negative       2.057455   \n",
       "                           antigen receptor-mediated signaling pathway (GO... IL-1β (Ca, 24h)  trained immunity True     Positive       2.455115   \n",
       "\n",
       "                                                                                                                                           P-value  \\\n",
       "Library                    Gene set                                           Cytokine         Phenotype        selected Association                 \n",
       "KEGG_2019_Human            Systemic lupus erythematosus                       IL-1β (Sa, 24h)  trained immunity True     Negative     3.301606e-07   \n",
       "GO_Biological_Process_2018 neutrophil degranulation (GO:0043312)              IL-2RA (Ca, 24h) trained immunity True     Negative     1.127527e-07   \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-2RA (Ca, 24h) trained immunity True     Negative     1.600209e-07   \n",
       "                           neutrophil activation involved in immune respon... IL-2RA (Ca, 24h) trained immunity True     Negative     1.681391e-07   \n",
       "                           antigen receptor-mediated signaling pathway (GO... IL-1β (Ca, 24h)  trained immunity True     Positive     5.868685e-07   \n",
       "\n",
       "                                                                                                                                      Adjusted P-value  \\\n",
       "Library                    Gene set                                           Cytokine         Phenotype        selected Association                     \n",
       "KEGG_2019_Human            Systemic lupus erythematosus                       IL-1β (Sa, 24h)  trained immunity True     Negative             0.000088   \n",
       "GO_Biological_Process_2018 neutrophil degranulation (GO:0043312)              IL-2RA (Ca, 24h) trained immunity True     Negative             0.000124   \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-2RA (Ca, 24h) trained immunity True     Negative             0.000124   \n",
       "                           neutrophil activation involved in immune respon... IL-2RA (Ca, 24h) trained immunity True     Negative             0.000124   \n",
       "                           antigen receptor-mediated signaling pathway (GO... IL-1β (Ca, 24h)  trained immunity True     Positive             0.001272   \n",
       "\n",
       "                                                                                                                                      Overlap  \n",
       "Library                    Gene set                                           Cytokine         Phenotype        selected Association           \n",
       "KEGG_2019_Human            Systemic lupus erythematosus                       IL-1β (Sa, 24h)  trained immunity True     Negative      17/160  \n",
       "GO_Biological_Process_2018 neutrophil degranulation (GO:0043312)              IL-2RA (Ca, 24h) trained immunity True     Negative     69/1474  \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-2RA (Ca, 24h) trained immunity True     Negative     69/1488  \n",
       "                           neutrophil activation involved in immune respon... IL-2RA (Ca, 24h) trained immunity True     Negative     69/1490  \n",
       "                           antigen receptor-mediated signaling pathway (GO... IL-1β (Ca, 24h)  trained immunity True     Positive      42/759  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library</th>\n",
       "      <th>Gene set</th>\n",
       "      <th>Cytokine</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th>Association</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">GO_Biological_Process_2018</th>\n",
       "      <th>neutrophil activation involved in immune response (GO:0002283)</th>\n",
       "      <th>IL-17 (Ca, 7d)</th>\n",
       "      <th>heterologous immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.847237</td>\n",
       "      <td>1.756606e-13</td>\n",
       "      <td>3.822375e-10</td>\n",
       "      <td>79/430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neutrophil degranulation (GO:0043312)</th>\n",
       "      <th>IL-17 (Ca, 7d)</th>\n",
       "      <th>heterologous immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.785488</td>\n",
       "      <td>8.631027e-13</td>\n",
       "      <td>9.390557e-10</td>\n",
       "      <td>77/426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neutrophil mediated immunity (GO:0002446)</th>\n",
       "      <th>IL-17 (Ca, 7d)</th>\n",
       "      <th>heterologous immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.729077</td>\n",
       "      <td>2.036914e-12</td>\n",
       "      <td>1.477442e-09</td>\n",
       "      <td>77/433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_2019_Human</th>\n",
       "      <th>Staphylococcus aureus infection</th>\n",
       "      <th>IL-17 (Ca, 7d)</th>\n",
       "      <th>heterologous immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>12.955366</td>\n",
       "      <td>3.624189e-11</td>\n",
       "      <td>9.749068e-09</td>\n",
       "      <td>17/33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO_Biological_Process_2018</th>\n",
       "      <th>regulation of immune response (GO:0050776)</th>\n",
       "      <th>IL-17 (Ca, 7d)</th>\n",
       "      <th>heterologous immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>4.444730</td>\n",
       "      <td>2.512261e-11</td>\n",
       "      <td>1.366670e-08</td>\n",
       "      <td>36/136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                         Odds ratio  \\\n",
       "Library                    Gene set                                           Cytokine       Phenotype             selected Association               \n",
       "GO_Biological_Process_2018 neutrophil activation involved in immune respon... IL-17 (Ca, 7d) heterologous immunity True     Positive       2.847237   \n",
       "                           neutrophil degranulation (GO:0043312)              IL-17 (Ca, 7d) heterologous immunity True     Positive       2.785488   \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-17 (Ca, 7d) heterologous immunity True     Positive       2.729077   \n",
       "KEGG_2019_Human            Staphylococcus aureus infection                    IL-17 (Ca, 7d) heterologous immunity True     Positive      12.955366   \n",
       "GO_Biological_Process_2018 regulation of immune response (GO:0050776)         IL-17 (Ca, 7d) heterologous immunity True     Positive       4.444730   \n",
       "\n",
       "                                                                                                                                              P-value  \\\n",
       "Library                    Gene set                                           Cytokine       Phenotype             selected Association                 \n",
       "GO_Biological_Process_2018 neutrophil activation involved in immune respon... IL-17 (Ca, 7d) heterologous immunity True     Positive     1.756606e-13   \n",
       "                           neutrophil degranulation (GO:0043312)              IL-17 (Ca, 7d) heterologous immunity True     Positive     8.631027e-13   \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-17 (Ca, 7d) heterologous immunity True     Positive     2.036914e-12   \n",
       "KEGG_2019_Human            Staphylococcus aureus infection                    IL-17 (Ca, 7d) heterologous immunity True     Positive     3.624189e-11   \n",
       "GO_Biological_Process_2018 regulation of immune response (GO:0050776)         IL-17 (Ca, 7d) heterologous immunity True     Positive     2.512261e-11   \n",
       "\n",
       "                                                                                                                                         Adjusted P-value  \\\n",
       "Library                    Gene set                                           Cytokine       Phenotype             selected Association                     \n",
       "GO_Biological_Process_2018 neutrophil activation involved in immune respon... IL-17 (Ca, 7d) heterologous immunity True     Positive         3.822375e-10   \n",
       "                           neutrophil degranulation (GO:0043312)              IL-17 (Ca, 7d) heterologous immunity True     Positive         9.390557e-10   \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-17 (Ca, 7d) heterologous immunity True     Positive         1.477442e-09   \n",
       "KEGG_2019_Human            Staphylococcus aureus infection                    IL-17 (Ca, 7d) heterologous immunity True     Positive         9.749068e-09   \n",
       "GO_Biological_Process_2018 regulation of immune response (GO:0050776)         IL-17 (Ca, 7d) heterologous immunity True     Positive         1.366670e-08   \n",
       "\n",
       "                                                                                                                                        Overlap  \n",
       "Library                    Gene set                                           Cytokine       Phenotype             selected Association          \n",
       "GO_Biological_Process_2018 neutrophil activation involved in immune respon... IL-17 (Ca, 7d) heterologous immunity True     Positive     79/430  \n",
       "                           neutrophil degranulation (GO:0043312)              IL-17 (Ca, 7d) heterologous immunity True     Positive     77/426  \n",
       "                           neutrophil mediated immunity (GO:0002446)          IL-17 (Ca, 7d) heterologous immunity True     Positive     77/433  \n",
       "KEGG_2019_Human            Staphylococcus aureus infection                    IL-17 (Ca, 7d) heterologous immunity True     Positive      17/33  \n",
       "GO_Biological_Process_2018 regulation of immune response (GO:0050776)         IL-17 (Ca, 7d) heterologous immunity True     Positive     36/136  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library</th>\n",
       "      <th>Region set</th>\n",
       "      <th>Cytokine</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th>Association</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Roadmap_Epigenomics_r9</th>\n",
       "      <th>H3K4me3</th>\n",
       "      <th>IL-10 (Mt, 24h)</th>\n",
       "      <th>innate after Mt stimulus</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>5.130075</td>\n",
       "      <td>7.576059e-132</td>\n",
       "      <td>5.303241e-131</td>\n",
       "      <td>742/23396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K27ac</th>\n",
       "      <th>IL-10 (Mt, 24h)</th>\n",
       "      <th>innate after Mt stimulus</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>3.883915</td>\n",
       "      <td>5.729741e-96</td>\n",
       "      <td>2.005409e-95</td>\n",
       "      <td>680/22982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">HOCOMOCO_v11</th>\n",
       "      <th>KAISO</th>\n",
       "      <th>IFN-γ (Mt, 7d)</th>\n",
       "      <th>adaptive immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>3.915944</td>\n",
       "      <td>4.658174e-89</td>\n",
       "      <td>1.867928e-86</td>\n",
       "      <td>759/28870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZBT14</th>\n",
       "      <th>IFN-γ (Mt, 7d)</th>\n",
       "      <th>adaptive immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>3.738352</td>\n",
       "      <td>6.950583e-88</td>\n",
       "      <td>1.393592e-85</td>\n",
       "      <td>717/26192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Roadmap_Epigenomics_r9</th>\n",
       "      <th>H3K4me3</th>\n",
       "      <th>IFN-γ (Mt, 7d)</th>\n",
       "      <th>adaptive immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>3.554554</td>\n",
       "      <td>1.084240e-84</td>\n",
       "      <td>7.589683e-84</td>\n",
       "      <td>667/23396</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                 Odds ratio  \\\n",
       "Library                Region set Cytokine        Phenotype                selected Association               \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-10 (Mt, 24h) innate after Mt stimulus True     Negative       5.130075   \n",
       "                       H3K27ac    IL-10 (Mt, 24h) innate after Mt stimulus True     Negative       3.883915   \n",
       "HOCOMOCO_v11           KAISO      IFN-γ (Mt, 7d)  adaptive immunity        True     Negative       3.915944   \n",
       "                       ZBT14      IFN-γ (Mt, 7d)  adaptive immunity        True     Negative       3.738352   \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IFN-γ (Mt, 7d)  adaptive immunity        True     Negative       3.554554   \n",
       "\n",
       "                                                                                                       P-value  \\\n",
       "Library                Region set Cytokine        Phenotype                selected Association                  \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-10 (Mt, 24h) innate after Mt stimulus True     Negative     7.576059e-132   \n",
       "                       H3K27ac    IL-10 (Mt, 24h) innate after Mt stimulus True     Negative      5.729741e-96   \n",
       "HOCOMOCO_v11           KAISO      IFN-γ (Mt, 7d)  adaptive immunity        True     Negative      4.658174e-89   \n",
       "                       ZBT14      IFN-γ (Mt, 7d)  adaptive immunity        True     Negative      6.950583e-88   \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IFN-γ (Mt, 7d)  adaptive immunity        True     Negative      1.084240e-84   \n",
       "\n",
       "                                                                                                 Adjusted P-value  \\\n",
       "Library                Region set Cytokine        Phenotype                selected Association                     \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-10 (Mt, 24h) innate after Mt stimulus True     Negative        5.303241e-131   \n",
       "                       H3K27ac    IL-10 (Mt, 24h) innate after Mt stimulus True     Negative         2.005409e-95   \n",
       "HOCOMOCO_v11           KAISO      IFN-γ (Mt, 7d)  adaptive immunity        True     Negative         1.867928e-86   \n",
       "                       ZBT14      IFN-γ (Mt, 7d)  adaptive immunity        True     Negative         1.393592e-85   \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IFN-γ (Mt, 7d)  adaptive immunity        True     Negative         7.589683e-84   \n",
       "\n",
       "                                                                                                   Overlap  \n",
       "Library                Region set Cytokine        Phenotype                selected Association             \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-10 (Mt, 24h) innate after Mt stimulus True     Negative     742/23396  \n",
       "                       H3K27ac    IL-10 (Mt, 24h) innate after Mt stimulus True     Negative     680/22982  \n",
       "HOCOMOCO_v11           KAISO      IFN-γ (Mt, 7d)  adaptive immunity        True     Negative     759/28870  \n",
       "                       ZBT14      IFN-γ (Mt, 7d)  adaptive immunity        True     Negative     717/26192  \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IFN-γ (Mt, 7d)  adaptive immunity        True     Negative     667/23396  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adjusting per phenotype\n",
      "adjusting per phenotype\n",
      "S03_ATACseq_d0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Fraction of variance explained (FVE) by donor ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Association of chromatin accessibility with ag...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Association of chromatin accessibility with se...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Association of chromatin accessibility with BM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Association of chromatin accessibility with or...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Association of chromatin accessibility with al...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Association of chromatin accessibility with sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Association of chromatin accessibility with in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d0</th>\n",
       "      <td>Association of chromatin accessibility with cy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_TRIM</th>\n",
       "      <td>Association of chromatin accessibility with BC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_ADAPT</th>\n",
       "      <td>Association of chromatin accessibility with BC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Fraction of variance explained (FVE) by donor ...\n",
       "d0_age                   Association of chromatin accessibility with ag...\n",
       "d0_sex                   Association of chromatin accessibility with se...\n",
       "d0_BMI                   Association of chromatin accessibility with BM...\n",
       "d0_oralContraceptives    Association of chromatin accessibility with or...\n",
       "d0_alcohol               Association of chromatin accessibility with al...\n",
       "d0_sampleCollectionTime  Association of chromatin accessibility with sa...\n",
       "d0_inflaMark_d0          Association of chromatin accessibility with in...\n",
       "d0_cytokine_d0           Association of chromatin accessibility with cy...\n",
       "d0_cytokine_d14FC_TRIM   Association of chromatin accessibility with BC...\n",
       "d0_cytokine_d14FC_ADAPT  Association of chromatin accessibility with BC..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S04_ATACseq_d0_geneSets\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_TRIM</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_ADAPT</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Chromatin accessibility gene set enrichment of...\n",
       "d0_age                   Chromatin accessibility gene set enrichment of...\n",
       "d0_sex                   Chromatin accessibility gene set enrichment of...\n",
       "d0_BMI                   Chromatin accessibility gene set enrichment of...\n",
       "d0_oralContraceptives    Chromatin accessibility gene set enrichment of...\n",
       "d0_alcohol               Chromatin accessibility gene set enrichment of...\n",
       "d0_sampleCollectionTime  Chromatin accessibility gene set enrichment of...\n",
       "d0_inflaMark_d0          Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d0           Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d14FC_TRIM   Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d14FC_ADAPT  Chromatin accessibility gene set enrichment of..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ext_S04.1_ATACseq_d0_geneSets_promoters\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_TRIM</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_ADAPT</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Chromatin accessibility gene set enrichment of...\n",
       "d0_age                   Chromatin accessibility gene set enrichment of...\n",
       "d0_sex                   Chromatin accessibility gene set enrichment of...\n",
       "d0_BMI                   Chromatin accessibility gene set enrichment of...\n",
       "d0_oralContraceptives    Chromatin accessibility gene set enrichment of...\n",
       "d0_alcohol               Chromatin accessibility gene set enrichment of...\n",
       "d0_sampleCollectionTime  Chromatin accessibility gene set enrichment of...\n",
       "d0_inflaMark_d0          Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d0           Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d14FC_TRIM   Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d14FC_ADAPT  Chromatin accessibility gene set enrichment of..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ext_S04.2_ATACseq_d0_regionSets\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d0</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_TRIM</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_ADAPT</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Chromatin accessibility region set enrichment ...\n",
       "d0_age                   Chromatin accessibility region set enrichment ...\n",
       "d0_sex                   Chromatin accessibility region set enrichment ...\n",
       "d0_BMI                   Chromatin accessibility region set enrichment ...\n",
       "d0_oralContraceptives    Chromatin accessibility region set enrichment ...\n",
       "d0_alcohol               Chromatin accessibility region set enrichment ...\n",
       "d0_sampleCollectionTime  Chromatin accessibility region set enrichment ...\n",
       "d0_inflaMark_d0          Chromatin accessibility region set enrichment ...\n",
       "d0_cytokine_d0           Chromatin accessibility region set enrichment ...\n",
       "d0_cytokine_d14FC_TRIM   Chromatin accessibility region set enrichment ...\n",
       "d0_cytokine_d14FC_ADAPT  Chromatin accessibility region set enrichment ..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "V3 CYTO\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Coefficient</th>\n",
       "      <th>T-value</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Promoter mapping</th>\n",
       "      <th>Distal mapping</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Region</th>\n",
       "      <th>Cytokine</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr20_59120306_59121000</th>\n",
       "      <th>IL-2RA (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <td>-0.600066</td>\n",
       "      <td>-5.268007</td>\n",
       "      <td>3.086457e-07</td>\n",
       "      <td>0.019773</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr19_991255_992114</th>\n",
       "      <th>TNF-α (LPS, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <td>-0.227095</td>\n",
       "      <td>-5.108948</td>\n",
       "      <td>6.910734e-07</td>\n",
       "      <td>0.044274</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WDR18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3_11633594_11634703</th>\n",
       "      <th>IL-1Ra (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <td>0.511821</td>\n",
       "      <td>5.044510</td>\n",
       "      <td>9.017168e-07</td>\n",
       "      <td>0.057768</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VGLL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr11_74004698_74005760</th>\n",
       "      <th>GM-CSF (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <td>-0.129874</td>\n",
       "      <td>-4.958361</td>\n",
       "      <td>1.354858e-06</td>\n",
       "      <td>0.086799</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UCP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr18_68735539_68736283</th>\n",
       "      <th>CCL4 (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <td>0.404066</td>\n",
       "      <td>4.777234</td>\n",
       "      <td>3.108496e-06</td>\n",
       "      <td>0.144981</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDC102B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                    Coefficient  \\\n",
       "Region                  Cytokine         Phenotype        selected                \n",
       "chr20_59120306_59121000 IL-2RA (Ca, 24h) trained immunity True        -0.600066   \n",
       "chr19_991255_992114     TNF-α (LPS, 24h) trained immunity True        -0.227095   \n",
       "chr3_11633594_11634703  IL-1Ra (Ca, 24h) trained immunity True         0.511821   \n",
       "chr11_74004698_74005760 GM-CSF (Ca, 24h) trained immunity True        -0.129874   \n",
       "chr18_68735539_68736283 CCL4 (Ca, 24h)   trained immunity True         0.404066   \n",
       "\n",
       "                                                                     T-value  \\\n",
       "Region                  Cytokine         Phenotype        selected             \n",
       "chr20_59120306_59121000 IL-2RA (Ca, 24h) trained immunity True     -5.268007   \n",
       "chr19_991255_992114     TNF-α (LPS, 24h) trained immunity True     -5.108948   \n",
       "chr3_11633594_11634703  IL-1Ra (Ca, 24h) trained immunity True      5.044510   \n",
       "chr11_74004698_74005760 GM-CSF (Ca, 24h) trained immunity True     -4.958361   \n",
       "chr18_68735539_68736283 CCL4 (Ca, 24h)   trained immunity True      4.777234   \n",
       "\n",
       "                                                                         P-value  \\\n",
       "Region                  Cytokine         Phenotype        selected                 \n",
       "chr20_59120306_59121000 IL-2RA (Ca, 24h) trained immunity True      3.086457e-07   \n",
       "chr19_991255_992114     TNF-α (LPS, 24h) trained immunity True      6.910734e-07   \n",
       "chr3_11633594_11634703  IL-1Ra (Ca, 24h) trained immunity True      9.017168e-07   \n",
       "chr11_74004698_74005760 GM-CSF (Ca, 24h) trained immunity True      1.354858e-06   \n",
       "chr18_68735539_68736283 CCL4 (Ca, 24h)   trained immunity True      3.108496e-06   \n",
       "\n",
       "                                                                    Adjusted P-value  \\\n",
       "Region                  Cytokine         Phenotype        selected                     \n",
       "chr20_59120306_59121000 IL-2RA (Ca, 24h) trained immunity True              0.019773   \n",
       "chr19_991255_992114     TNF-α (LPS, 24h) trained immunity True              0.044274   \n",
       "chr3_11633594_11634703  IL-1Ra (Ca, 24h) trained immunity True              0.057768   \n",
       "chr11_74004698_74005760 GM-CSF (Ca, 24h) trained immunity True              0.086799   \n",
       "chr18_68735539_68736283 CCL4 (Ca, 24h)   trained immunity True              0.144981   \n",
       "\n",
       "                                                                   Promoter mapping  \\\n",
       "Region                  Cytokine         Phenotype        selected                    \n",
       "chr20_59120306_59121000 IL-2RA (Ca, 24h) trained immunity True                  NaN   \n",
       "chr19_991255_992114     TNF-α (LPS, 24h) trained immunity True                  NaN   \n",
       "chr3_11633594_11634703  IL-1Ra (Ca, 24h) trained immunity True                  NaN   \n",
       "chr11_74004698_74005760 GM-CSF (Ca, 24h) trained immunity True                  NaN   \n",
       "chr18_68735539_68736283 CCL4 (Ca, 24h)   trained immunity True                  NaN   \n",
       "\n",
       "                                                                   Distal mapping  \n",
       "Region                  Cytokine         Phenotype        selected                 \n",
       "chr20_59120306_59121000 IL-2RA (Ca, 24h) trained immunity True             ZNF831  \n",
       "chr19_991255_992114     TNF-α (LPS, 24h) trained immunity True              WDR18  \n",
       "chr3_11633594_11634703  IL-1Ra (Ca, 24h) trained immunity True              VGLL4  \n",
       "chr11_74004698_74005760 GM-CSF (Ca, 24h) trained immunity True               UCP3  \n",
       "chr18_68735539_68736283 CCL4 (Ca, 24h)   trained immunity True           CCDC102B  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library</th>\n",
       "      <th>Gene set</th>\n",
       "      <th>Cytokine</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th>Association</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO_Biological_Process_2018</th>\n",
       "      <th>translation (GO:0006412)</th>\n",
       "      <th>IL-1Ra (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>3.454319</td>\n",
       "      <td>8.589877e-08</td>\n",
       "      <td>0.000187</td>\n",
       "      <td>28/366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">KEGG_2019_Human</th>\n",
       "      <th>VEGF signaling pathway</th>\n",
       "      <th>IFN-γ (Sa, 7d)</th>\n",
       "      <th>heterologous immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>3.611274</td>\n",
       "      <td>9.647378e-07</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>22/275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Natural killer cell mediated cytotoxicity</th>\n",
       "      <th>IFN-γ (Sa, 7d)</th>\n",
       "      <th>heterologous immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>2.553918</td>\n",
       "      <td>1.018124e-05</td>\n",
       "      <td>0.001339</td>\n",
       "      <td>30/519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">GO_Biological_Process_2018</th>\n",
       "      <th>positive regulation of interleukin-2 production (GO:0032743)</th>\n",
       "      <th>TNF-α (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>4.542631</td>\n",
       "      <td>9.213781e-07</td>\n",
       "      <td>0.001605</td>\n",
       "      <td>17/172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>antigen receptor-mediated signaling pathway (GO:0050851)</th>\n",
       "      <th>TNF-α (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.390767</td>\n",
       "      <td>1.461185e-06</td>\n",
       "      <td>0.001605</td>\n",
       "      <td>41/759</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                           Odds ratio  \\\n",
       "Library                    Gene set                                           Cytokine         Phenotype             selected Association               \n",
       "GO_Biological_Process_2018 translation (GO:0006412)                           IL-1Ra (Ca, 24h) trained immunity      True     Positive       3.454319   \n",
       "KEGG_2019_Human            VEGF signaling pathway                             IFN-γ (Sa, 7d)   heterologous immunity True     Negative       3.611274   \n",
       "                           Natural killer cell mediated cytotoxicity          IFN-γ (Sa, 7d)   heterologous immunity True     Negative       2.553918   \n",
       "GO_Biological_Process_2018 positive regulation of interleukin-2 production... TNF-α (Ca, 24h)  trained immunity      True     Positive       4.542631   \n",
       "                           antigen receptor-mediated signaling pathway (GO... TNF-α (Ca, 24h)  trained immunity      True     Positive       2.390767   \n",
       "\n",
       "                                                                                                                                                P-value  \\\n",
       "Library                    Gene set                                           Cytokine         Phenotype             selected Association                 \n",
       "GO_Biological_Process_2018 translation (GO:0006412)                           IL-1Ra (Ca, 24h) trained immunity      True     Positive     8.589877e-08   \n",
       "KEGG_2019_Human            VEGF signaling pathway                             IFN-γ (Sa, 7d)   heterologous immunity True     Negative     9.647378e-07   \n",
       "                           Natural killer cell mediated cytotoxicity          IFN-γ (Sa, 7d)   heterologous immunity True     Negative     1.018124e-05   \n",
       "GO_Biological_Process_2018 positive regulation of interleukin-2 production... TNF-α (Ca, 24h)  trained immunity      True     Positive     9.213781e-07   \n",
       "                           antigen receptor-mediated signaling pathway (GO... TNF-α (Ca, 24h)  trained immunity      True     Positive     1.461185e-06   \n",
       "\n",
       "                                                                                                                                           Adjusted P-value  \\\n",
       "Library                    Gene set                                           Cytokine         Phenotype             selected Association                     \n",
       "GO_Biological_Process_2018 translation (GO:0006412)                           IL-1Ra (Ca, 24h) trained immunity      True     Positive             0.000187   \n",
       "KEGG_2019_Human            VEGF signaling pathway                             IFN-γ (Sa, 7d)   heterologous immunity True     Negative             0.000254   \n",
       "                           Natural killer cell mediated cytotoxicity          IFN-γ (Sa, 7d)   heterologous immunity True     Negative             0.001339   \n",
       "GO_Biological_Process_2018 positive regulation of interleukin-2 production... TNF-α (Ca, 24h)  trained immunity      True     Positive             0.001605   \n",
       "                           antigen receptor-mediated signaling pathway (GO... TNF-α (Ca, 24h)  trained immunity      True     Positive             0.001605   \n",
       "\n",
       "                                                                                                                                          Overlap  \n",
       "Library                    Gene set                                           Cytokine         Phenotype             selected Association          \n",
       "GO_Biological_Process_2018 translation (GO:0006412)                           IL-1Ra (Ca, 24h) trained immunity      True     Positive     28/366  \n",
       "KEGG_2019_Human            VEGF signaling pathway                             IFN-γ (Sa, 7d)   heterologous immunity True     Negative     22/275  \n",
       "                           Natural killer cell mediated cytotoxicity          IFN-γ (Sa, 7d)   heterologous immunity True     Negative     30/519  \n",
       "GO_Biological_Process_2018 positive regulation of interleukin-2 production... TNF-α (Ca, 24h)  trained immunity      True     Positive     17/172  \n",
       "                           antigen receptor-mediated signaling pathway (GO... TNF-α (Ca, 24h)  trained immunity      True     Positive     41/759  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library</th>\n",
       "      <th>Gene set</th>\n",
       "      <th>Cytokine</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th>Association</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">GO_Biological_Process_2018</th>\n",
       "      <th>neutrophil activation involved in immune response (GO:0002283)</th>\n",
       "      <th>TNF-α (LPS, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.799929</td>\n",
       "      <td>5.025179e-13</td>\n",
       "      <td>1.137198e-09</td>\n",
       "      <td>78/430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neutrophil degranulation (GO:0043312)</th>\n",
       "      <th>TNF-α (LPS, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.738256</td>\n",
       "      <td>2.416639e-12</td>\n",
       "      <td>2.734427e-09</td>\n",
       "      <td>76/426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neutrophil mediated immunity (GO:0002446)</th>\n",
       "      <th>TNF-α (LPS, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.682952</td>\n",
       "      <td>5.589475e-12</td>\n",
       "      <td>4.216327e-09</td>\n",
       "      <td>76/433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>response to lipopolysaccharide (GO:0032496)</th>\n",
       "      <th>TNF-α (LPS, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>3.987952</td>\n",
       "      <td>8.904228e-10</td>\n",
       "      <td>5.037567e-07</td>\n",
       "      <td>34/139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_2019_Human</th>\n",
       "      <th>Legionellosis</th>\n",
       "      <th>IL-1β (LPS, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>6.452138</td>\n",
       "      <td>2.638873e-08</td>\n",
       "      <td>4.532072e-06</td>\n",
       "      <td>18/52</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                      Odds ratio  \\\n",
       "Library                    Gene set                                           Cytokine         Phenotype        selected Association               \n",
       "GO_Biological_Process_2018 neutrophil activation involved in immune respon... TNF-α (LPS, 24h) trained immunity True     Positive       2.799929   \n",
       "                           neutrophil degranulation (GO:0043312)              TNF-α (LPS, 24h) trained immunity True     Positive       2.738256   \n",
       "                           neutrophil mediated immunity (GO:0002446)          TNF-α (LPS, 24h) trained immunity True     Positive       2.682952   \n",
       "                           response to lipopolysaccharide (GO:0032496)        TNF-α (LPS, 24h) trained immunity True     Positive       3.987952   \n",
       "KEGG_2019_Human            Legionellosis                                      IL-1β (LPS, 24h) trained immunity True     Positive       6.452138   \n",
       "\n",
       "                                                                                                                                           P-value  \\\n",
       "Library                    Gene set                                           Cytokine         Phenotype        selected Association                 \n",
       "GO_Biological_Process_2018 neutrophil activation involved in immune respon... TNF-α (LPS, 24h) trained immunity True     Positive     5.025179e-13   \n",
       "                           neutrophil degranulation (GO:0043312)              TNF-α (LPS, 24h) trained immunity True     Positive     2.416639e-12   \n",
       "                           neutrophil mediated immunity (GO:0002446)          TNF-α (LPS, 24h) trained immunity True     Positive     5.589475e-12   \n",
       "                           response to lipopolysaccharide (GO:0032496)        TNF-α (LPS, 24h) trained immunity True     Positive     8.904228e-10   \n",
       "KEGG_2019_Human            Legionellosis                                      IL-1β (LPS, 24h) trained immunity True     Positive     2.638873e-08   \n",
       "\n",
       "                                                                                                                                      Adjusted P-value  \\\n",
       "Library                    Gene set                                           Cytokine         Phenotype        selected Association                     \n",
       "GO_Biological_Process_2018 neutrophil activation involved in immune respon... TNF-α (LPS, 24h) trained immunity True     Positive         1.137198e-09   \n",
       "                           neutrophil degranulation (GO:0043312)              TNF-α (LPS, 24h) trained immunity True     Positive         2.734427e-09   \n",
       "                           neutrophil mediated immunity (GO:0002446)          TNF-α (LPS, 24h) trained immunity True     Positive         4.216327e-09   \n",
       "                           response to lipopolysaccharide (GO:0032496)        TNF-α (LPS, 24h) trained immunity True     Positive         5.037567e-07   \n",
       "KEGG_2019_Human            Legionellosis                                      IL-1β (LPS, 24h) trained immunity True     Positive         4.532072e-06   \n",
       "\n",
       "                                                                                                                                     Overlap  \n",
       "Library                    Gene set                                           Cytokine         Phenotype        selected Association          \n",
       "GO_Biological_Process_2018 neutrophil activation involved in immune respon... TNF-α (LPS, 24h) trained immunity True     Positive     78/430  \n",
       "                           neutrophil degranulation (GO:0043312)              TNF-α (LPS, 24h) trained immunity True     Positive     76/426  \n",
       "                           neutrophil mediated immunity (GO:0002446)          TNF-α (LPS, 24h) trained immunity True     Positive     76/433  \n",
       "                           response to lipopolysaccharide (GO:0032496)        TNF-α (LPS, 24h) trained immunity True     Positive     34/139  \n",
       "KEGG_2019_Human            Legionellosis                                      IL-1β (LPS, 24h) trained immunity True     Positive      18/52  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Odds ratio</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library</th>\n",
       "      <th>Region set</th>\n",
       "      <th>Cytokine</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>selected</th>\n",
       "      <th>Association</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Roadmap_Epigenomics_r9</th>\n",
       "      <th>H3K4me3</th>\n",
       "      <th>IL-1Ra (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.861328</td>\n",
       "      <td>5.272896e-60</td>\n",
       "      <td>3.691027e-59</td>\n",
       "      <td>618/23396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K9me3</th>\n",
       "      <th>TNF-α (LPS, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>6.845209</td>\n",
       "      <td>1.049417e-57</td>\n",
       "      <td>7.345916e-57</td>\n",
       "      <td>130/1477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">HOCOMOCO_v11</th>\n",
       "      <th>CEBPB</th>\n",
       "      <th>TNF-α (LPS, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>3.227006</td>\n",
       "      <td>2.936488e-51</td>\n",
       "      <td>1.177532e-48</td>\n",
       "      <td>845/40460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBD2</th>\n",
       "      <th>TNF-α (LPS, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Negative</th>\n",
       "      <td>2.753859</td>\n",
       "      <td>4.101403e-50</td>\n",
       "      <td>1.644663e-47</td>\n",
       "      <td>740/32793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Roadmap_Epigenomics_r9</th>\n",
       "      <th>H3K27ac</th>\n",
       "      <th>IL-1Ra (Ca, 24h)</th>\n",
       "      <th>trained immunity</th>\n",
       "      <th>True</th>\n",
       "      <th>Positive</th>\n",
       "      <td>2.527639</td>\n",
       "      <td>1.832831e-47</td>\n",
       "      <td>6.414909e-47</td>\n",
       "      <td>582/22982</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                          Odds ratio  \\\n",
       "Library                Region set Cytokine         Phenotype        selected Association               \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-1Ra (Ca, 24h) trained immunity True     Positive       2.861328   \n",
       "                       H3K9me3    TNF-α (LPS, 24h) trained immunity True     Negative       6.845209   \n",
       "HOCOMOCO_v11           CEBPB      TNF-α (LPS, 24h) trained immunity True     Positive       3.227006   \n",
       "                       MBD2       TNF-α (LPS, 24h) trained immunity True     Negative       2.753859   \n",
       "Roadmap_Epigenomics_r9 H3K27ac    IL-1Ra (Ca, 24h) trained immunity True     Positive       2.527639   \n",
       "\n",
       "                                                                                               P-value  \\\n",
       "Library                Region set Cytokine         Phenotype        selected Association                 \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-1Ra (Ca, 24h) trained immunity True     Positive     5.272896e-60   \n",
       "                       H3K9me3    TNF-α (LPS, 24h) trained immunity True     Negative     1.049417e-57   \n",
       "HOCOMOCO_v11           CEBPB      TNF-α (LPS, 24h) trained immunity True     Positive     2.936488e-51   \n",
       "                       MBD2       TNF-α (LPS, 24h) trained immunity True     Negative     4.101403e-50   \n",
       "Roadmap_Epigenomics_r9 H3K27ac    IL-1Ra (Ca, 24h) trained immunity True     Positive     1.832831e-47   \n",
       "\n",
       "                                                                                          Adjusted P-value  \\\n",
       "Library                Region set Cytokine         Phenotype        selected Association                     \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-1Ra (Ca, 24h) trained immunity True     Positive         3.691027e-59   \n",
       "                       H3K9me3    TNF-α (LPS, 24h) trained immunity True     Negative         7.345916e-57   \n",
       "HOCOMOCO_v11           CEBPB      TNF-α (LPS, 24h) trained immunity True     Positive         1.177532e-48   \n",
       "                       MBD2       TNF-α (LPS, 24h) trained immunity True     Negative         1.644663e-47   \n",
       "Roadmap_Epigenomics_r9 H3K27ac    IL-1Ra (Ca, 24h) trained immunity True     Positive         6.414909e-47   \n",
       "\n",
       "                                                                                            Overlap  \n",
       "Library                Region set Cytokine         Phenotype        selected Association             \n",
       "Roadmap_Epigenomics_r9 H3K4me3    IL-1Ra (Ca, 24h) trained immunity True     Positive     618/23396  \n",
       "                       H3K9me3    TNF-α (LPS, 24h) trained immunity True     Negative      130/1477  \n",
       "HOCOMOCO_v11           CEBPB      TNF-α (LPS, 24h) trained immunity True     Positive     845/40460  \n",
       "                       MBD2       TNF-α (LPS, 24h) trained immunity True     Negative     740/32793  \n",
       "Roadmap_Epigenomics_r9 H3K27ac    IL-1Ra (Ca, 24h) trained immunity True     Positive     582/22982  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adjusting per phenotype\n",
      "adjusting per phenotype\n",
      "S03_ATACseq_d0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Fraction of variance explained (FVE) by donor ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Association of chromatin accessibility with ag...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Association of chromatin accessibility with se...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Association of chromatin accessibility with BM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Association of chromatin accessibility with or...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Association of chromatin accessibility with al...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Association of chromatin accessibility with sa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Association of chromatin accessibility with in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d0</th>\n",
       "      <td>Association of chromatin accessibility with cy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_TRIM</th>\n",
       "      <td>Association of chromatin accessibility with BC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_ADAPT</th>\n",
       "      <td>Association of chromatin accessibility with BC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d90FC_TRIM</th>\n",
       "      <td>Association of chromatin accessibility with BC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d90FC_ADAPT</th>\n",
       "      <td>Association of chromatin accessibility with BC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Fraction of variance explained (FVE) by donor ...\n",
       "d0_age                   Association of chromatin accessibility with ag...\n",
       "d0_sex                   Association of chromatin accessibility with se...\n",
       "d0_BMI                   Association of chromatin accessibility with BM...\n",
       "d0_oralContraceptives    Association of chromatin accessibility with or...\n",
       "d0_alcohol               Association of chromatin accessibility with al...\n",
       "d0_sampleCollectionTime  Association of chromatin accessibility with sa...\n",
       "d0_inflaMark_d0          Association of chromatin accessibility with in...\n",
       "d0_cytokine_d0           Association of chromatin accessibility with cy...\n",
       "d0_cytokine_d14FC_TRIM   Association of chromatin accessibility with BC...\n",
       "d0_cytokine_d14FC_ADAPT  Association of chromatin accessibility with BC...\n",
       "d0_cytokine_d90FC_TRIM   Association of chromatin accessibility with BC...\n",
       "d0_cytokine_d90FC_ADAPT  Association of chromatin accessibility with BC..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S04_ATACseq_d0_geneSets\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_TRIM</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_ADAPT</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d90FC_TRIM</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d90FC_ADAPT</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Chromatin accessibility gene set enrichment of...\n",
       "d0_age                   Chromatin accessibility gene set enrichment of...\n",
       "d0_sex                   Chromatin accessibility gene set enrichment of...\n",
       "d0_BMI                   Chromatin accessibility gene set enrichment of...\n",
       "d0_oralContraceptives    Chromatin accessibility gene set enrichment of...\n",
       "d0_alcohol               Chromatin accessibility gene set enrichment of...\n",
       "d0_sampleCollectionTime  Chromatin accessibility gene set enrichment of...\n",
       "d0_inflaMark_d0          Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d0           Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d14FC_TRIM   Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d14FC_ADAPT  Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d90FC_TRIM   Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d90FC_ADAPT  Chromatin accessibility gene set enrichment of..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ext_S04.1_ATACseq_d0_geneSets_promoters\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d0</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_TRIM</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_ADAPT</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d90FC_TRIM</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d90FC_ADAPT</th>\n",
       "      <td>Chromatin accessibility gene set enrichment of...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Chromatin accessibility gene set enrichment of...\n",
       "d0_age                   Chromatin accessibility gene set enrichment of...\n",
       "d0_sex                   Chromatin accessibility gene set enrichment of...\n",
       "d0_BMI                   Chromatin accessibility gene set enrichment of...\n",
       "d0_oralContraceptives    Chromatin accessibility gene set enrichment of...\n",
       "d0_alcohol               Chromatin accessibility gene set enrichment of...\n",
       "d0_sampleCollectionTime  Chromatin accessibility gene set enrichment of...\n",
       "d0_inflaMark_d0          Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d0           Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d14FC_TRIM   Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d14FC_ADAPT  Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d90FC_TRIM   Chromatin accessibility gene set enrichment of...\n",
       "d0_cytokine_d90FC_ADAPT  Chromatin accessibility gene set enrichment of..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ext_S04.2_ATACseq_d0_regionSets\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sheet name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>donorVar</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_age</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sex</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_BMI</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_oralContraceptives</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_alcohol</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_sampleCollectionTime</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_inflaMark_d0</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d0</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_TRIM</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d14FC_ADAPT</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d90FC_TRIM</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d0_cytokine_d90FC_ADAPT</th>\n",
       "      <td>Chromatin accessibility region set enrichment ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Contents\n",
       "Sheet name                                                                \n",
       "donorVar                 Chromatin accessibility region set enrichment ...\n",
       "d0_age                   Chromatin accessibility region set enrichment ...\n",
       "d0_sex                   Chromatin accessibility region set enrichment ...\n",
       "d0_BMI                   Chromatin accessibility region set enrichment ...\n",
       "d0_oralContraceptives    Chromatin accessibility region set enrichment ...\n",
       "d0_alcohol               Chromatin accessibility region set enrichment ...\n",
       "d0_sampleCollectionTime  Chromatin accessibility region set enrichment ...\n",
       "d0_inflaMark_d0          Chromatin accessibility region set enrichment ...\n",
       "d0_cytokine_d0           Chromatin accessibility region set enrichment ...\n",
       "d0_cytokine_d14FC_TRIM   Chromatin accessibility region set enrichment ...\n",
       "d0_cytokine_d14FC_ADAPT  Chromatin accessibility region set enrichment ...\n",
       "d0_cytokine_d90FC_TRIM   Chromatin accessibility region set enrichment ...\n",
       "d0_cytokine_d90FC_ADAPT  Chromatin accessibility region set enrichment ..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df = misc.get_sample_annot()\n",
    "peaks_df = misc.get_peak_annot()\n",
    "table_name = ATAC_V1\n",
    "for visit, Y_name, de_fig_n, enr_fig_n in [\n",
    "    ('V1', 'CM', '3D', '3E'),\n",
    "    ('V1', 'CYTO', '4E', '4F'),\n",
    "    ('V2', 'CYTO', None, None),\n",
    "    ('V3', 'CYTO', '6D', '6E')\n",
    "]:\n",
    "    # START\n",
    "    suppl_tables = misc.get_container_with_tables(\n",
    "        [ATAC_V1, ENRICH_V1, ENRICH_V1_PROMOTERS, ENRICH_V1_LOLA],\n",
    "        modes=['a', 'a', 'a', 'a']\n",
    "    )\n",
    "    \n",
    "    print(visit, Y_name)\n",
    "    if Y_name == 'CYTO':\n",
    "        proteins = df.columns[(df.columns.str.contains('^CYTO:.*_good$') & ~df.columns.str.contains('lactate'))]\n",
    "    else:\n",
    "        proteins = df.columns[df.columns.str.contains('^CM:')]\n",
    "\n",
    "    if DEBUG:\n",
    "        proteins = proteins[[-2, -7, -6]]\n",
    "        \n",
    "    de_suppl_dict = {}\n",
    "    enr_suppl_dict = defaultdict(lambda: defaultdict(lambda: defaultdict(lambda: defaultdict(lambda: {}))))\n",
    "    \n",
    "    for protein in proteins:\n",
    "        coef = f'{f\"LFC_{visit}_CORR_\" if visit != \"V1\" else \"\"}{misc.safe_R_name(protein)}'\n",
    "        model = f'V1.batch.sex.age{\"\" if Y_name == \"CYTO\" else \".bmi.oralContra\"}.blood.TSS_enr.visit_time.{coef}'\n",
    "        de_suppl_dict[protein] = misc.read_de(CELLTYPE, model, contrasts=coef)\n",
    "        \n",
    "        for region_filter in [misc.GENE_AND_DISTAL_10kb, misc.TSS_PROXIMAL]:\n",
    "            for direction in [-1, 1]:\n",
    "                enr_fn = misc.enrichr_fn(CELLTYPE, model, data='results', coef=coef, effect_size_filter=None, rank_metric=RANK_METRIC,\n",
    "                                         top_n=TOP_N, direction=direction, db=None, regions=region_filter)\n",
    "                enr_df = pd.read_csv(enr_fn, sep='\\t', usecols=['Gene_set'] + get_ENR_suppl_cols())\n",
    "                for library in [misc.GO_BIO_PROCESS, misc.KEGG]:\n",
    "                    enr_suppl_dict['enrichr'][region_filter][library][direction][protein] = enr_df.loc[enr_df['Gene_set'] == library + '_min15_max500'].copy()\n",
    "        \n",
    "        for direction in [-1, 1]:\n",
    "            enr_df = misc.read_enrichr_or_lola(CELLTYPE, model, coef=coef, effect_size_filter=None,\n",
    "                                               rank_metric=RANK_METRIC, top_n=TOP_N, region_filter=None,\n",
    "                                               method='lola', direction=direction)\n",
    "            for library in [HOCOMOCO, EPI_ROADMAP]:\n",
    "                library_mask = enr_df['Gene_set'] == library\n",
    "                enr_df.loc[library_mask, 'Adjusted P-value'] = misc.adjusted_pvals(enr_df.loc[library_mask, 'P-value'], method='fdr_bh')\n",
    "                enr_suppl_dict['lola'][None][library][direction][protein] = enr_df.loc[library_mask].reset_index().copy()\n",
    "        del enr_df\n",
    "        \n",
    "    de_suppl_df = misc.make_suppl_table_atac_seq_d0_immune(de_suppl_dict, Y_name, proteins, visit,\n",
    "                                                           peaks_df=peaks_df, region_filter=misc.GENE_AND_DISTAL_10kb,\n",
    "                                                           limit_top_n=TOP_N, rank_metric=RANK_METRIC)\n",
    "    del de_suppl_dict\n",
    "    display(de_suppl_df.head())\n",
    "    \n",
    "    enr_suppl_dfs = defaultdict(lambda: {})\n",
    "    for method in enr_suppl_dict:\n",
    "        for region_filter in enr_suppl_dict[method]:\n",
    "            enr_suppl_dfs[method][region_filter] = misc.make_suppl_table_enrich_d0_immune(enr_suppl_dict[method][region_filter], Y_name, proteins,\n",
    "                                                                                          method, limit_pval=ENR_PVAL)\n",
    "            display(enr_suppl_dfs[method][region_filter].head())\n",
    "    del enr_suppl_dict\n",
    "    \n",
    "    cytokine_fold_changes = visit != 'V1' and Y_name == 'CYTO'\n",
    "    # save selected regions to supplement; if cytokine_fold_changes, split by phenotype and subset to phenotype-related cytokines\n",
    "    misc.suppl_table_atac_seq_d0_immune(de_suppl_df.loc[~de_suppl_df[DISTAL_MAP_COL].isnull()], Y_name, visit,\n",
    "                                        split_by_phenotype=cytokine_fold_changes, split_by_protein=False,\n",
    "                                        only_selected=True, phenotypes=['innate_nonspecific_24h_wo_LAC', 'adaptive_MTB_7d'],\n",
    "                                        drop_padj=False, suppl_tables=suppl_tables, gene_sets=False, table_name=table_name, fig_n=de_fig_n)\n",
    "    # save all gene sets to supplement; if cytokine_fold_changes, split by phenotype and subset to phenotype-related cytokines\n",
    "    for method, region_filter, enr_table_name, fig_n in [\n",
    "        ('enrichr', misc.GENE_AND_DISTAL_10kb, ENRICH_V1, enr_fig_n),\n",
    "        ('enrichr', misc.TSS_PROXIMAL, ENRICH_V1_PROMOTERS, None),\n",
    "        ('lola', None, ENRICH_V1_LOLA, None)\n",
    "    ]: \n",
    "        misc.suppl_table_atac_seq_d0_immune(enr_suppl_dfs[method][region_filter], Y_name, visit,\n",
    "                                            split_by_phenotype=cytokine_fold_changes, split_by_protein=False,\n",
    "                                            only_selected=True, phenotypes=['innate_nonspecific_24h_wo_LAC', 'adaptive_MTB_7d'],\n",
    "                                            promoters=region_filter == misc.TSS_PROXIMAL,\n",
    "                                            method=method, drop_padj=True,\n",
    "                                            suppl_tables=suppl_tables, gene_sets=True, table_name=enr_table_name, fig_n=fig_n)\n",
    "    \n",
    "    if SAVE_EXT_TABLES:\n",
    "        for protein in proteins:\n",
    "            protein = misc.rename_Y_vars(Y_name, protein, with_time=True)\n",
    "            print(protein)\n",
    "            protein_wo_spaces = protein.replace('(', '').replace(')', '').replace(\n",
    "                ',', '').replace(' ', '_').replace('α', 'a').replace('β', 'b').replace('γ', 'g')\n",
    "            ext_table_name = f'{protein_wo_spaces}_{NAMES[\"ATAC\"]}_d0_{NAMES[Y_name]}_{misc.get_day_suffix(visit)}'\n",
    "            ext_suppl_tables = misc.get_container_with_tables(ext_table_name, include_legends_sheet=False)\n",
    "            for sheet_name, method, region_filter in [\n",
    "                (None, None, None),\n",
    "                ('geneSets', 'enrichr', misc.GENE_AND_DISTAL_10kb),\n",
    "                ('geneSets_promoters', 'enrichr', misc.TSS_PROXIMAL),\n",
    "                ('regionSets', 'lola', None)\n",
    "            ]:\n",
    "                if sheet_name is None:\n",
    "                    assert method is None and region_filter is None\n",
    "                    suppl_df = de_suppl_df\n",
    "                else:\n",
    "                    suppl_df = enr_suppl_dfs[method][region_filter]\n",
    "                # select proteins\n",
    "                suppl_df = suppl_df.loc[suppl_df.index.get_level_values(LONG_NAMES[Y_name]) == protein]\n",
    "\n",
    "                misc.suppl_table_atac_seq_d0_immune(suppl_df, Y_name, visit,\n",
    "                                                    split_by_phenotype=False, split_by_protein=False, only_selected=False,\n",
    "                                                    promoters=region_filter == misc.TSS_PROXIMAL,\n",
    "                                                    method=method, gene_sets=method is not None, drop_padj=True, force_sheet_name=sheet_name,\n",
    "                                                    suppl_tables=ext_suppl_tables, table_name=ext_table_name, fig_n=de_fig_n[0] if de_fig_n else None, phenotypes=None)\n",
    "\n",
    "        ext_suppl_tables[ext_table_name]['file'].close()\n",
    "        misc.fix_font_for_suppl_tables(ext_suppl_tables, fix_legends_sheet=False)\n",
    "        \n",
    "    del enr_suppl_dfs\n",
    "    del de_suppl_df\n",
    "    \n",
    "    # FINISH\n",
    "    infos, tables = misc.close_tables_and_save(suppl_tables)\n",
    "    for table, info_df in zip(tables, infos):\n",
    "        print(table)\n",
    "        display(info_df)  \n",
    "    misc.fix_font_for_suppl_tables(suppl_tables)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BCG (Python 3)",
   "language": "python",
   "name": "bcg_python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
